import{_ as e}from"./plugin-vue_export-helper-DlAUqK2U.js";import{c as t,o as a,d as r}from"./app-D-JbdY3s.js";const n={},s=r(`<h1 id="the-e-utilities-in-depth-parameters-syntax-and-more" tabindex="-1"><a class="header-anchor" href="#the-e-utilities-in-depth-parameters-syntax-and-more"><span>The E-utilities In-Depth: Parameters, Syntax and More</span></a></h1><blockquote><p>Eric Sayers, PhD.</p><p><a href="https://www.ncbi.nlm.nih.gov/books/NBK25498/#__NBK25498_ai__" target="_blank" rel="noopener noreferrer">Author Information and Affiliations</a></p><p><strong>Authors</strong></p><p>Eric Sayers, PhD<img src="https://www.ncbi.nlm.nih.gov/corehtml/pmc/pmcgifs/corrauth.gif" alt="corresponding author" loading="lazy">1.</p><p><strong>Affilitions</strong></p><p>1 NCBI</p><p>Email: <a href="mailto:dev@null" target="_blank" rel="noopener noreferrer">vog.hin.mln.ibcn@sreyas</a></p><p><img src="https://www.ncbi.nlm.nih.gov/corehtml/pmc/pmcgifs/corrauth.gif" alt="corresponding author" loading="lazy">Corresponding author.</p><p>Created: May 29, 2009; Last Update: November 30, 2022.</p><p><em>Estimated reading time: 20 minutes</em></p></blockquote><h2 id="introduction" tabindex="-1"><a class="header-anchor" href="#introduction"><span>Introduction</span></a></h2><p>This chapter serves as a reference for all supported parameters for the E-utilities, along with accepted values and usage guidelines. This information is provided for each E-utility in sections below, and parameters and/or values specific to particular databases are discussed within each section. Most E-utilities have a set of parameters that are required for any call, in addition to several additional optional parameters that extend the tool&#39;s functionality. These two sets of parameters are discussed separately in each section.</p><h2 id="general-usage-guidelines" tabindex="-1"><a class="header-anchor" href="#general-usage-guidelines"><span>General Usage Guidelines</span></a></h2><p>Please see <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/" target="_blank" rel="noopener noreferrer">Chapter 2</a> for a detailed discussion of E-utility usage policy. The following two parameters should be included in all E-utility requests.</p><h3 id="tool" tabindex="-1"><a class="header-anchor" href="#tool"><span>tool</span></a></h3><p>Name of application making the E-utility call. Value must be a string with no internal spaces.</p><h3 id="email" tabindex="-1"><a class="header-anchor" href="#email"><span>email</span></a></h3><p>E-mail address of the E-utility user. Value must be a string with no internal spaces, and should be a valid e-mail address.</p><p>If you expect to post more than 3 E-utility requests per second from a single IP address, consider including the following parameter:</p><h3 id="api-key" tabindex="-1"><a class="header-anchor" href="#api-key"><span><strong>api_key</strong></span></a></h3><p>Value of the API key for sites that post more than 3 requests per second. Please see <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/" target="_blank" rel="noopener noreferrer">Chapter 2</a> for a full discussion of this policy.</p><h2 id="e-utilities-dtds" tabindex="-1"><a class="header-anchor" href="#e-utilities-dtds"><span>E-utilities DTDs</span></a></h2><p>With the exception of EFetch, the E-utilities each generate a single XML output format that conforms to a DTD specific for that utility. Links to these DTDs are provided in the XML headers of the E-utility returns.</p><p>ESummary version 2.0 produces unique XML DocSums for each Entrez database, and as such each Entrez database has a unique DTD for version 2.0 DocSums. Links to these DTDs are provided in the version 2.0 XML.</p><p>EFetch produces output in a variety of formats, some of which are XML. Most of these XML formats also conform to DTDs or schema specific to the relevant Entrez database. Please follow the appropriate link below for the PubMed DTD:</p><ul><li><a href="http://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_180601.dtd" target="_blank" rel="noopener noreferrer">PubMed DTD June 2018 – current PubMed DTD</a></li><li><a href="http://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_190101.dtd" target="_blank" rel="noopener noreferrer">PubMed DTD January 2019 – forthcoming DTD</a></li></ul><h2 id="einfo" tabindex="-1"><a class="header-anchor" href="#einfo"><span>EInfo</span></a></h2><h3 id="base-url" tabindex="-1"><a class="header-anchor" href="#base-url"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi</p><h3 id="functions" tabindex="-1"><a class="header-anchor" href="#functions"><span>Functions</span></a></h3><ul><li>Provides a list of the names of all valid Entrez databases</li><li>Provides statistics for a single database, including lists of indexing fields and available link names</li></ul><h3 id="required-parameters" tabindex="-1"><a class="header-anchor" href="#required-parameters"><span>Required Parameters</span></a></h3><p>None. If no <strong>db</strong> parameter is provided, einfo will return a list of the names of all valid Entrez databases.</p><h3 id="optional-parameters" tabindex="-1"><a class="header-anchor" href="#optional-parameters"><span>Optional Parameters</span></a></h3><h4 id="db" tabindex="-1"><a class="header-anchor" href="#db"><span>db</span></a></h4><p>Target database about which to gather statistics. Value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a>.</p><h4 id="version" tabindex="-1"><a class="header-anchor" href="#version"><span>version</span></a></h4><p>Used to specify version 2.0 EInfo XML. The only supported value is ‘2.0’. When present, EInfo will return XML that includes two new fields: <code>&lt;IsTruncatable&gt;</code> and <code>&lt;IsRangeable&gt;</code>. Fields that are truncatable allow the wildcard character ‘*’ in terms. The wildcard character will expand to match any set of characters up to a limit of 600 unique expansions. Fields that are rangeable allow the range operator ‘:’ to be placed between a lower and upper limit for the desired range (e.g. 2008:2010[pdat]).</p><h4 id="retmode" tabindex="-1"><a class="header-anchor" href="#retmode"><span>retmode</span></a></h4><p>Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for EInfo XML, but ‘json’ is also supported to return output in JSON format.</p><h3 id="examples" tabindex="-1"><a class="header-anchor" href="#examples"><span>Examples</span></a></h3><p>Return a list of all Entrez database names:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi</strong></p><p>Return version 2.0 statistics for Entrez Protein:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=protein&amp;version=2.0</strong></p><h2 id="esearch" tabindex="-1"><a class="header-anchor" href="#esearch"><span>ESearch</span></a></h2><h3 id="base-url-1" tabindex="-1"><a class="header-anchor" href="#base-url-1"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi</p><h3 id="functions-1" tabindex="-1"><a class="header-anchor" href="#functions-1"><span>Functions</span></a></h3><ul><li>Provides a list of UIDs matching a text query</li><li>Posts the results of a search on the History server</li><li>Downloads all UIDs from a dataset stored on the History server</li><li>Combines or limits UID datasets stored on the History server</li><li>Sorts sets of UIDs</li></ul><p>API users should be aware that some NCBI products contain search tools that generate content from searches on the web interface that are not available to ESearch. For example, the PubMed web interface (pubmed.ncbi.nlm.nih.gov) contains citation matching and spelling correction tools that are only available through that interface. Please see ECitMatch and ESpell below for API equivalents.</p><h3 id="required-parameters-1" tabindex="-1"><a class="header-anchor" href="#required-parameters-1"><span>Required Parameters</span></a></h3><h4 id="db-1" tabindex="-1"><a class="header-anchor" href="#db-1"><span>db</span></a></h4><p>Database to search. Value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a> (default = pubmed).</p><h4 id="term" tabindex="-1"><a class="header-anchor" href="#term"><span>term</span></a></h4><p>Entrez text query. All special characters must be URL encoded. Spaces may be replaced by &#39;+&#39; signs. For very long queries (more than several hundred characters long), consider using an HTTP POST call. See the <a href="https://pubmed.ncbi.nlm.nih.gov/help/" target="_blank" rel="noopener noreferrer">PubMed</a> or <a href="https://www.ncbi.nlm.nih.gov/books/n/helpentrez/EntrezHelp/" target="_blank" rel="noopener noreferrer">Entrez</a> help for information about search field descriptions and tags. Search fields and tags are database specific.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>esearch.fcgi?db=pubmed&amp;term=asthma</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>PubMed also offers “<a href="https://pubmed.ncbi.nlm.nih.gov/help/#proximity-searching" target="_blank" rel="noopener noreferrer">proximity searching</a>” for multiple terms appearing in any order within a specified number of words from one another in the [Title] or [Title/Abstract] fields.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>esearch.fcgi?db=pubmed&amp;term=”asthma treatment”[Title:~3]</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h3 id="optional-parameters-–-history-server" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-history-server"><span>Optional Parameters – History Server</span></a></h3><h4 id="usehistory" tabindex="-1"><a class="header-anchor" href="#usehistory"><span>usehistory</span></a></h4><p>When <strong>usehistory</strong> is set to &#39;y&#39;, ESearch will post the UIDs resulting from the search operation onto the History server so that they can be used directly in a subsequent E-utility call. Also, <strong>usehistory</strong> must be set to &#39;y&#39; for ESearch to interpret query key values included in <strong>term</strong> or to accept a <strong>WebEnv</strong> as input.</p><h4 id="webenv" tabindex="-1"><a class="header-anchor" href="#webenv"><span>WebEnv</span></a></h4><p>Web environment string returned from a previous ESearch, EPost or ELink call. When provided, ESearch will post the results of the search operation to this pre-existing WebEnv, thereby appending the results to the existing environment. In addition, providing <strong>WebEnv</strong> allows query keys to be used in <strong>term</strong> so that previous search sets can be combined or limited. As described above, if <strong>WebEnv</strong> is used, <strong>usehistory</strong> must be set to &#39;y&#39;.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>esearch.fcgi?db=pubmed&amp;term=asthma&amp;WebEnv=&lt;webenv string&gt;&amp;usehistory=y</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h4 id="query-key" tabindex="-1"><a class="header-anchor" href="#query-key"><span>query_key</span></a></h4><p>Integer query key returned by a previous ESearch, EPost or ELink call. When provided, ESearch will find the intersection of the set specified by <strong>query_key</strong> and the set retrieved by the query in <strong>term</strong> (i.e. joins the two with AND). For <strong>query_key</strong> to function, <strong>WebEnv</strong> must be assigned an existing WebEnv string and <strong>usehistory</strong> must be set to &#39;y&#39;.</p><p>Values for query keys may also be provided in <strong>term</strong> if they are preceeded by a &#39;#&#39; (%23 in the URL). While only one <strong>query_key</strong> parameter can be provided to ESearch, any number of query keys can be combined in <strong>term</strong>. Also, if query keys are provided in <strong>term</strong>, they can be combined with OR or NOT in addition to AND.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>The following two URLs are functionally equivalent:</span></span>
<span class="line"><span></span></span>
<span class="line"><span>esearch.fcgi?db=pubmed&amp;term=asthma&amp;query_key=1&amp;WebEnv=&lt;webenv string&gt;&amp;usehistory=y</span></span>
<span class="line"><span></span></span>
<span class="line"><span>esearch.fcgi?db=pubmed&amp;term=%231+AND+asthma&amp;WebEnv=&lt;webenv string&gt;&amp;usehistory=y</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="optional-parameters-–-retrieval" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-retrieval"><span>Optional Parameters – Retrieval</span></a></h3><h4 id="retstart" tabindex="-1"><a class="header-anchor" href="#retstart"><span>retstart</span></a></h4><p>Sequential index of the first UID in the retrieved set to be shown in the XML output (default=0, corresponding to the first record of the entire set). This parameter can be used in conjunction with <strong>retmax</strong> to download an arbitrary subset of UIDs retrieved from a search.</p><h4 id="retmax" tabindex="-1"><a class="header-anchor" href="#retmax"><span>retmax</span></a></h4><p>Total number of UIDs from the retrieved set to be shown in the XML output (default=20). By default, ESearch only includes the first 20 UIDs retrieved in the XML output. If <strong>usehistory</strong> is set to &#39;y&#39;, the remainder of the retrieved set will be stored on the History server; otherwise these UIDs are lost. Increasing <strong>retmax</strong> allows more of the retrieved UIDs to be included in the XML output, up to a maximum of 10,000 records.</p><p>To retrieve more than 10,000 UIDs from databases other than PubMed, submit multiple esearch requests while incrementing the value of <strong>retstart</strong> (see Application 3). For PubMed, ESearch can only retrieve the first 10,000 records matching the query. To obtain more than 10,000 PubMed records, consider using <code>&lt;EDirect&gt;</code> that contains additional logic to batch PubMed search results automatically so that an arbitrary number can be retrieved.</p><h4 id="rettype" tabindex="-1"><a class="header-anchor" href="#rettype"><span>rettype</span></a></h4><p>Retrieval type. There are two allowed values for ESearch: &#39;uilist&#39; (default), which displays the standard XML output, and &#39;count&#39;, which displays only the <code>&lt;Count&gt;</code> tag.</p><h4 id="retmode-1" tabindex="-1"><a class="header-anchor" href="#retmode-1"><span>retmode</span></a></h4><p>Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for ESearch XML, but ‘json’ is also supported to return output in JSON format.</p><h4 id="sort" tabindex="-1"><a class="header-anchor" href="#sort"><span>sort</span></a></h4><p>Specifies the method used to sort UIDs in the ESearch output. The available values vary by database (<strong>db</strong>) and may be found in the Display Settings menu on an Entrez search results page. If <strong>usehistory</strong> is set to ‘y’, the UIDs are loaded onto the History Server in the specified sort order and will be retrieved in that order by ESummary or EFetch. Example values are ‘relevance’ and ‘name’ for Gene. Users should be aware that the default value of <strong>sort</strong> varies from one database to another, and that the default value used by ESearch for a given database may differ from that used on NCBI web search pages.</p><p>Values of <strong>sort</strong> for PubMed are as follows:</p><ul><li><em>pub_date</em> – descending sort by publication date</li><li><em>Author</em> – ascending sort by first author</li><li><em>JournalName</em> – ascending sort by journal name</li><li><em>relevance</em> – default sort order, (“Best Match”) on web PubMed</li></ul><h4 id="field" tabindex="-1"><a class="header-anchor" href="#field"><span>field</span></a></h4><p>Search field. If used, the entire search term will be limited to the specified Entrez field. The following two URLs are equivalent:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>esearch.fcgi?db=pubmed&amp;term=asthma&amp;field=title</span></span>
<span class="line"><span></span></span>
<span class="line"><span>esearch.fcgi?db=pubmed&amp;term=asthma[title]</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h4 id="idtype" tabindex="-1"><a class="header-anchor" href="#idtype"><span>idtype</span></a></h4><p>Specifies the type of identifier to return for sequence databases (nuccore, popset, protein). By default, ESearch returns GI numbers in its output. If <strong>idtype</strong> is set to ‘acc’, ESearch will return accession.version identifiers rather than GI numbers.</p><h3 id="optional-parameters-–-dates" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-dates"><span>Optional Parameters – Dates</span></a></h3><h4 id="datetype" tabindex="-1"><a class="header-anchor" href="#datetype"><span>datetype</span></a></h4><p>Type of date used to limit a search. The allowed values vary between Entrez databases, but common values are &#39;mdat&#39; (modification date), &#39;pdat&#39; (publication date) and &#39;edat&#39; (Entrez date). Generally an Entrez database will have only two allowed values for <strong>datetype</strong>.</p><h4 id="reldate" tabindex="-1"><a class="header-anchor" href="#reldate"><span>reldate</span></a></h4><p>When <strong>reldate</strong> is set to an integer <em>n</em>, the search returns only those items that have a date specified by <strong>datetype</strong> within the last <em>n</em> days.</p><h4 id="mindate-maxdate" tabindex="-1"><a class="header-anchor" href="#mindate-maxdate"><span>mindate, maxdate</span></a></h4><p>Date range used to limit a search result by the date specified by <strong>datetype</strong>. These two parameters (<strong>mindate, maxdate</strong>) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM.</p><h3 id="examples-1" tabindex="-1"><a class="header-anchor" href="#examples-1"><span>Examples</span></a></h3><p>Search in PubMed with the term <em>cancer</em> for abstracts that have an Entrez date within the last 60 days; retrieve the first 100 PMIDs and translations; post the results on the History server and return a <strong>WebEnv</strong> and <strong>query_key</strong>:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=cancer&amp;reldate=60&amp;datetype=edat&amp;retmax=100&amp;usehistory=y</strong></p><p>Search in PubMed for the journal PNAS, Volume 97, and retrieve six PMIDs starting with the seventh PMID in the list:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=PNAS[ta]+AND+97[vi]&amp;retstart=6&amp;retmax=6&amp;tool=biomed3</strong></p><p>Search in the NLM Catalog for journals matching the term <em>obstetrics</em>:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nlmcatalog&amp;term=obstetrics+AND+ncbijournals[filter]</strong></p><p>Search PubMed Central for free full text articles containing the query <em>stem cells</em>:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pmc&amp;term=stem+cells+AND+free+fulltext[filter]</strong></p><p>Search in Nucleotide for all tRNAs:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&amp;term=biomol+trna[prop]</strong></p><p>Search in Protein for a molecular weight range:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=protein&amp;term=70000:90000[molecular+weight]</strong></p><h2 id="epost" tabindex="-1"><a class="header-anchor" href="#epost"><span>EPost</span></a></h2><h3 id="base-url-2" tabindex="-1"><a class="header-anchor" href="#base-url-2"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi</p><h3 id="functions-2" tabindex="-1"><a class="header-anchor" href="#functions-2"><span>Functions</span></a></h3><ul><li>Uploads a list of UIDs to the Entrez History server</li><li>Appends a list of UIDs to an existing set of UID lists attached to a Web Environment</li></ul><h3 id="required-parameters-2" tabindex="-1"><a class="header-anchor" href="#required-parameters-2"><span>Required Parameters</span></a></h3><h4 id="db-2" tabindex="-1"><a class="header-anchor" href="#db-2"><span>db</span></a></h4><p>Database containing the UIDs in the input list. The value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a> (default = pubmed).</p><h4 id="id" tabindex="-1"><a class="header-anchor" href="#id"><span>id</span></a></h4><p>UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by <strong>db</strong>. For PubMed, no more than 10,000 UIDs can be included in a single URL request. For other databases there is no set maximum for the number of UIDs that can be passed to epost, but if more than about 200 UIDs are to be posted, the request should be made using the HTTP POST method.</p><p>For sequence databases (nuccore, popset, protein), the UID list may be a mixed list of GI numbers and accession.version identifiers. <strong>Note:</strong> When using accession.version identifiers, there is a conversion step that takes place that causes large lists of identifiers to time out, even when using POST. Therefore, we recommend batching these types of requests in sizes of about 500 UIDs or less, to avoid retrieving only a partial amount of records from your original POST input list.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>epost.fcgi?db=pubmed&amp;id=19393038,30242208,29453458</span></span>
<span class="line"><span>epost.fcgi?db=protein&amp;id=15718680,NP_001098858.1,119703751</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="optional-parameter" tabindex="-1"><a class="header-anchor" href="#optional-parameter"><span>Optional Parameter</span></a></h3><h4 id="webenv-1" tabindex="-1"><a class="header-anchor" href="#webenv-1"><span>WebEnv</span></a></h4><p>Web Environment. If provided, this parameter specifies the Web Environment that will receive the UID list sent by post. EPost will create a new query key associated with that Web Environment. Usually this WebEnv value is obtained from the output of a previous ESearch, EPost or ELink call. If no <strong>WebEnv</strong> parameter is provided, EPost will create a new Web Environment and post the UID list to <strong>query_key</strong> 1.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>epost.fcgi?db=protein&amp;id=15718680,157427902,119703751&amp;WebEnv=&lt;webenv string&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h3 id="example" tabindex="-1"><a class="header-anchor" href="#example"><span>Example</span></a></h3><p>Post records to PubMed:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi?db=pubmed&amp;id=11237011,12466850</strong></p><h2 id="esummary" tabindex="-1"><a class="header-anchor" href="#esummary"><span>ESummary</span></a></h2><h3 id="base-url-3" tabindex="-1"><a class="header-anchor" href="#base-url-3"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi</p><h3 id="functions-3" tabindex="-1"><a class="header-anchor" href="#functions-3"><span>Functions</span></a></h3><ul><li>Returns document summaries (DocSums) for a list of input UIDs</li><li>Returns DocSums for a set of UIDs stored on the Entrez History server</li></ul><h3 id="required-parameter" tabindex="-1"><a class="header-anchor" href="#required-parameter"><span>Required Parameter</span></a></h3><h4 id="db-3" tabindex="-1"><a class="header-anchor" href="#db-3"><span>db</span></a></h4><p>Database from which to retrieve DocSums. The value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a> (default = pubmed).</p><h3 id="required-parameter-–-used-only-when-input-is-from-a-uid-list" tabindex="-1"><a class="header-anchor" href="#required-parameter-–-used-only-when-input-is-from-a-uid-list"><span>Required Parameter – Used only when input is from a UID list</span></a></h3><h4 id="id-1" tabindex="-1"><a class="header-anchor" href="#id-1"><span>id</span></a></h4><p>UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by <strong>db</strong>. There is no set maximum for the number of UIDs that can be passed to ESummary, but if more than about 200 UIDs are to be provided, the request should be made using the HTTP POST method.</p><p>For sequence databases (nuccore, popset, protein), the UID list may be a mixed list of GI numbers and accession.version identifiers.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>esummary.fcgi?db=pubmed&amp;id=19393038,30242208,29453458</span></span>
<span class="line"><span>esummary.fcgi?db=protein&amp;id=15718680,NP_001098858.1,119703751</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="required-parameters-–-used-only-when-input-is-from-the-entrez-history-server" tabindex="-1"><a class="header-anchor" href="#required-parameters-–-used-only-when-input-is-from-the-entrez-history-server"><span>Required Parameters – Used only when input is from the Entrez History server</span></a></h3><h4 id="query-key-1" tabindex="-1"><a class="header-anchor" href="#query-key-1"><span>query_key</span></a></h4><p>Query key. This integer specifies which of the UID lists attached to the given Web Environment will be used as input to ESummary. Query keys are obtained from the output of previous ESearch, EPost or ELink calls. The <strong>query_key</strong> parameter must be used in conjunction with <strong>WebEnv</strong>.</p><h4 id="webenv-2" tabindex="-1"><a class="header-anchor" href="#webenv-2"><span>WebEnv</span></a></h4><p>Web Environment. This parameter specifies the Web Environment that contains the UID list to be provided as input to ESummary. Usually this WebEnv value is obtained from the output of a previous ESearch, EPost or ELink call. The <strong>WebEnv</strong> parameter must be used in conjunction with <strong>query_key</strong>.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>esummary.fcgi?db=protein&amp;query_key=&lt;key&gt;&amp;WebEnv=&lt;webenv string&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h3 id="optional-parameters-–-retrieval-1" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-retrieval-1"><span>Optional Parameters – Retrieval</span></a></h3><h4 id="retstart-1" tabindex="-1"><a class="header-anchor" href="#retstart-1"><span>retstart</span></a></h4><p>Sequential index of the first DocSum to be retrieved (default=1, corresponding to the first record of the entire set). This parameter can be used in conjunction with <strong>retmax</strong> to download an arbitrary subset of DocSums from the input set.</p><h4 id="retmax-1" tabindex="-1"><a class="header-anchor" href="#retmax-1"><span>retmax</span></a></h4><p>Total number of DocSums from the input set to be retrieved, up to a maximum of 10,000. If the total set is larger than this maximum, the value of <strong>retstart</strong> can be iterated while holding <strong>retmax</strong> constant, thereby downloading the entire set in batches of size <strong>retmax</strong>.</p><h4 id="retmode-2" tabindex="-1"><a class="header-anchor" href="#retmode-2"><span>retmode</span></a></h4><p>Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for ESummary XML, but ‘json’ is also supported to return output in JSON format.</p><h4 id="version-1" tabindex="-1"><a class="header-anchor" href="#version-1"><span>version</span></a></h4><p>Used to specify version 2.0 ESummary XML. The only supported value is ‘2.0’. When present, ESummary will return version 2.0 DocSum XML that is unique to each Entrez database and that often contains more data than the default DocSum XML.</p><h3 id="examples-2" tabindex="-1"><a class="header-anchor" href="#examples-2"><span>Examples</span></a></h3><p>PubMed:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&amp;id=11850928,11482001</strong></p><p>PubMed, version 2.0 XML:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&amp;id=11850928,11482001&amp;version=2.0</strong></p><p>Protein:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=protein&amp;id=28800982,28628843</strong></p><p>Nucleotide:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=nucleotide&amp;id=28864546,28800981</strong></p><p>Structure:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=structure&amp;id=19923,12120</strong></p><p>Taxonomy:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=taxonomy&amp;id=9913,30521</strong></p><p>UniSTS:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=unists&amp;id=254085,254086</strong></p><h2 id="efetch" tabindex="-1"><a class="header-anchor" href="#efetch"><span>EFetch</span></a></h2><h3 id="base-url-4" tabindex="-1"><a class="header-anchor" href="#base-url-4"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi</p><h3 id="functions-4" tabindex="-1"><a class="header-anchor" href="#functions-4"><span>Functions</span></a></h3><ul><li>Returns formatted data records for a list of input UIDs</li><li>Returns formatted data records for a set of UIDs stored on the Entrez History server</li></ul><h3 id="required-parameters-3" tabindex="-1"><a class="header-anchor" href="#required-parameters-3"><span>Required Parameters</span></a></h3><h4 id="db-4" tabindex="-1"><a class="header-anchor" href="#db-4"><span>db</span></a></h4><p>Database from which to retrieve records. The value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a> (default = pubmed). Currently EFetch does not support all Entrez databases. Please see <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.T._entrez_unique_identifiers_ui" target="_blank" rel="noopener noreferrer">Table 1</a> in Chapter 2 for a list of available databases.</p><h3 id="required-parameter-–-used-only-when-input-is-from-a-uid-list-1" tabindex="-1"><a class="header-anchor" href="#required-parameter-–-used-only-when-input-is-from-a-uid-list-1"><span>Required Parameter – Used only when input is from a UID list</span></a></h3><h4 id="id-2" tabindex="-1"><a class="header-anchor" href="#id-2"><span>id</span></a></h4><p>UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by <strong>db</strong>. There is no set maximum for the number of UIDs that can be passed to EFetch, but if more than about 200 UIDs are to be provided, the request should be made using the HTTP POST method.</p><p>For sequence databases (nuccore, popset, protein), the UID list may be a mixed list of GI numbers and accession.version identifiers.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>efetch.fcgi?db=pubmed&amp;id=19393038,30242208,29453458</span></span>
<span class="line"><span>efetch.fcgi?db=protein&amp;id=15718680,NP_001098858.1,119703751</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p><em>Special note for sequence databases.</em></p><p>NCBI is no longer assigning GI numbers to a growing number of new sequence records. As such, these records are not indexed in Entrez, and so cannot be retrieved using ESearch or ESummary, and have no Entrez links accessible by ELink. EFetch <em>can</em> retrieve these records by including their accession.version identifier in the <strong>id</strong> parameter.</p><h3 id="required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-1" tabindex="-1"><a class="header-anchor" href="#required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-1"><span>Required Parameters – Used only when input is from the Entrez History server</span></a></h3><h4 id="query-key-2" tabindex="-1"><a class="header-anchor" href="#query-key-2"><span>query_key</span></a></h4><p>Query key. This integer specifies which of the UID lists attached to the given Web Environment will be used as input to EFetch. Query keys are obtained from the output of previous ESearch, EPost or ELInk calls. The <strong>query_key</strong> parameter must be used in conjunction with <strong>WebEnv</strong>.</p><h4 id="webenv-3" tabindex="-1"><a class="header-anchor" href="#webenv-3"><span>WebEnv</span></a></h4><p>Web Environment. This parameter specifies the Web Environment that contains the UID list to be provided as input to EFetch. Usually this WebEnv value is obtained from the output of a previous ESearch, EPost or ELink call. The <strong>WebEnv</strong> parameter must be used in conjunction with <strong>query_key</strong>.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>efetch.fcgi?db=protein&amp;query_key=&lt;key&gt;&amp;WebEnv=&lt;webenv string&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h3 id="optional-parameters-–-retrieval-2" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-retrieval-2"><span>Optional Parameters – Retrieval</span></a></h3><h4 id="retmode-3" tabindex="-1"><a class="header-anchor" href="#retmode-3"><span>retmode</span></a></h4><p>Retrieval mode. This parameter specifies the data format of the records returned, such as plain text, HMTL or XML. See <a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly" target="_blank" rel="noopener noreferrer">Table 1</a> for a full list of allowed values for each database.</p><figure><a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly" target="_blank" rel="noopener noreferrer"><img src="https://www.ncbi.nlm.nih.gov/corehtml/pmc/css/bookshelf/2.26/img/table-icon.gif" alt="Table Icon" tabindex="0" loading="lazy"></a><figcaption>Table Icon</figcaption></figure><h4 id="table-1" tabindex="-1"><a class="header-anchor" href="#table-1"><span><a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly" target="_blank" rel="noopener noreferrer">Table 1</a></span></a></h4><p>– Valid values of &amp;retmode and &amp;rettype for EFetch (null = empty string)</p><h4 id="rettype-1" tabindex="-1"><a class="header-anchor" href="#rettype-1"><span>rettype</span></a></h4><p>Retrieval type. This parameter specifies the record view returned, such as Abstract or MEDLINE from PubMed, or GenPept or FASTA from protein. Please see <a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly" target="_blank" rel="noopener noreferrer">Table 1</a> for a full list of allowed values for each database.</p><h4 id="retstart-2" tabindex="-1"><a class="header-anchor" href="#retstart-2"><span>retstart</span></a></h4><p>Sequential index of the first record to be retrieved (default=0, corresponding to the first record of the entire set). This parameter can be used in conjunction with <strong>retmax</strong> to download an arbitrary subset of records from the input set.</p><h4 id="retmax-2" tabindex="-1"><a class="header-anchor" href="#retmax-2"><span>retmax</span></a></h4><p>Total number of records from the input set to be retrieved, up to a maximum of 10,000. Optionally, for a large set the value of <strong>retstart</strong> can be iterated while holding <strong>retmax</strong> constant, thereby downloading the entire set in batches of size <strong>retmax</strong>.</p><h3 id="optional-parameters-–-sequence-databases" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-sequence-databases"><span>Optional Parameters – Sequence Databases</span></a></h3><h4 id="strand" tabindex="-1"><a class="header-anchor" href="#strand"><span>strand</span></a></h4><p>Strand of DNA to retrieve. Available values are &quot;1&quot; for the plus strand and &quot;2&quot; for the minus strand.</p><h4 id="seq-start" tabindex="-1"><a class="header-anchor" href="#seq-start"><span>seq_start</span></a></h4><p>First sequence base to retrieve. The value should be the integer coordinate of the first desired base, with &quot;1&quot; representing the first base of the seqence.</p><h4 id="seq-stop" tabindex="-1"><a class="header-anchor" href="#seq-stop"><span>seq_stop</span></a></h4><p>Last sequence base to retrieve. The value should be the integer coordinate of the last desired base, with &quot;1&quot; representing the first base of the seqence.</p><h4 id="complexity" tabindex="-1"><a class="header-anchor" href="#complexity"><span>complexity</span></a></h4><p>Data content to return. Many sequence records are part of a larger data structure or &quot;blob&quot;, and the <strong>complexity</strong> parameter determines how much of that blob to return. For example, an mRNA may be stored together with its protein product. The available values are as follows:</p><table><thead><tr><th style="text-align:left;">Value of complexity</th><th style="text-align:left;">Data returned for each requested GI</th></tr></thead><tbody><tr><td style="text-align:left;">0</td><td style="text-align:left;">entire blob</td></tr><tr><td style="text-align:left;">1</td><td style="text-align:left;">bioseq</td></tr><tr><td style="text-align:left;">2</td><td style="text-align:left;">minimal bioseq-set</td></tr><tr><td style="text-align:left;">3</td><td style="text-align:left;">minimal nuc-prot</td></tr><tr><td style="text-align:left;">4</td><td style="text-align:left;">minimal pub-set</td></tr></tbody></table><h3 id="examples-3" tabindex="-1"><a class="header-anchor" href="#examples-3"><span>Examples</span></a></h3><p><strong>PubMed</strong></p><p>Fetch PMIDs 17284678 and 9997 as text abstracts:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&amp;id=17284678,9997&amp;retmode=text&amp;rettype=abstract</strong></p><p>Fetch PMIDs in XML:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&amp;id=11748933,11700088&amp;retmode=xml</strong></p><p><strong>PubMed Central</strong></p><p>Fetch XML for PubMed Central ID 212403:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pmc&amp;id=212403</strong></p><p><strong>Nucleotide/Nuccore</strong></p><p>Fetch the first 100 bases of the plus strand of GI 21614549 in FASTA format:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&amp;id=21614549&amp;strand=1&amp;seq_start=1&amp;seq_stop=100&amp;rettype=fasta&amp;retmode=text</strong></p><p>Fetch the first 100 bases of the minus strand of GI 21614549 in FASTA format:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&amp;id=21614549&amp;strand=2&amp;seq_start=1&amp;seq_stop=100&amp;rettype=fasta&amp;retmode=text</strong></p><p>Fetch the nuc-prot object for GI 21614549:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&amp;id=21614549&amp;complexity=3</strong></p><p>Fetch the full ASN.1 record for GI 5:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&amp;id=5</strong></p><p>Fetch FASTA for GI 5:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&amp;id=5&amp;rettype=fasta</strong></p><p>Fetch the GenBank flat file for GI 5:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&amp;id=5&amp;rettype=gb</strong></p><p>Fetch GBSeqXML for GI 5:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&amp;id=5&amp;rettype=gb&amp;retmode=xml</strong></p><p>Fetch TinySeqXML for GI 5:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&amp;id=5&amp;rettype=fasta&amp;retmode=xml</strong></p><p><strong>Popset</strong></p><p>Fetch the GenPept flat file for Popset ID 12829836:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=popset&amp;id=12829836&amp;rettype=gp</strong></p><p><strong>Protein</strong></p><p>Fetch the GenPept flat file for GI 8:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&amp;id=8&amp;rettype=gp</strong></p><p>Fetch GBSeqXML for GI 8:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&amp;id=8&amp;rettype=gp&amp;retmode=xml</strong></p><p><strong>Sequences</strong></p><p>Fetch FASTA for a transcript and its protein product (GIs 312836839 and 34577063)</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=sequences&amp;id=312836839,34577063&amp;rettype=fasta&amp;retmode=text</strong></p><p><strong>Gene</strong></p><p>Fetch full XML record for Gene ID 2:</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&amp;id=2&amp;retmode=xml</strong></p><h2 id="elink" tabindex="-1"><a class="header-anchor" href="#elink"><span>ELink</span></a></h2><h3 id="base-url-5" tabindex="-1"><a class="header-anchor" href="#base-url-5"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi</p><h3 id="functions-5" tabindex="-1"><a class="header-anchor" href="#functions-5"><span>Functions</span></a></h3><ul><li>Returns UIDs linked to an input set of UIDs in either the same or a different Entrez database</li><li>Returns UIDs linked to other UIDs in the same Entrez database that match an Entrez query</li><li>Checks for the existence of Entrez links for a set of UIDs within the same database</li><li>Lists the available links for a UID</li><li>Lists LinkOut URLs and attributes for a set of UIDs</li><li>Lists hyperlinks to primary LinkOut providers for a set of UIDs</li><li>Creates hyperlinks to the primary LinkOut provider for a single UID</li></ul><h3 id="required-parameters-4" tabindex="-1"><a class="header-anchor" href="#required-parameters-4"><span>Required Parameters</span></a></h3><h4 id="db-5" tabindex="-1"><a class="header-anchor" href="#db-5"><span>db</span></a></h4><p>Database from which to retrieve UIDs. The value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a> (default = pubmed). This is the destination database for the link operation.</p><h4 id="dbfrom" tabindex="-1"><a class="header-anchor" href="#dbfrom"><span>dbfrom</span></a></h4><p>Database containing the input UIDs. The value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a> (default = pubmed). This is the origin database of the link operation. If <strong>db</strong> and <strong>dbfrom</strong> are set to the same database value, then ELink will return computational neighbors within that database. Please see the <a href="https://eutils.ncbi.nlm.nih.gov/entrez/query/static/entrezlinks.html" target="_blank" rel="noopener noreferrer">full list of Entrez links</a> for available computational neighbors. Computational neighbors have linknames that begin with <em>dbname_dbname</em> (examples: protein_protein, pcassay_pcassay_activityneighbor).</p><h4 id="cmd" tabindex="-1"><a class="header-anchor" href="#cmd"><span>cmd</span></a></h4><p>ELink command mode. The command mode specified which function ELink will perform. Some optional parameters only function for certain values of &amp;cmd (see below).</p><p><strong>cmd=neighbor (default)</strong></p><p>ELink returns a set of UIDs in <strong>db</strong> linked to the input UIDs in <strong>dbfrom.</strong></p><p><em>Example: Link from protein to gene</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;db=gene&amp;id=15718680,157427902</strong></p><p><strong>cmd=neighbor_score</strong></p><p>ELink returns a set of UIDs within the same database as the input UIDs along with computed similarity scores**.**</p><p><em>Example: Find related articles to PMID 20210808</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;db=pubmed&amp;id=20210808&amp;cmd=neighbor_score</strong></p><p><strong>cmd=neighbor_history</strong></p><p>ELink posts the output UIDs to the Entrez History server and returns a <strong>query_key</strong> and <strong>WebEnv</strong> corresponding to the location of the output set.</p><p><em>Example: Link from protein to gene and post the results on the Entrez History</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;db=gene&amp;id=15718680,157427902&amp;cmd=neighbor_history</strong></p><p><strong>cmd=acheck</strong></p><p>ELink lists all links available for a set of UIDs.</p><p><em>Example: List all possible links from two protein GIs</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;id=15718680,157427902&amp;cmd=acheck</strong></p><p><em>Example: List all possible links from two protein GIs to PubMed</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;db=pubmed&amp;id=15718680,157427902&amp;cmd=acheck</strong></p><p><strong>cmd=ncheck</strong></p><p>ELink checks for the existence of links <em>within the same database</em> for a set of UIDs. These links are equivalent to setting <strong>db</strong> and <strong>dbfrom</strong> to the same value.</p><p><em>Example: Check whether two nuccore sequences have &quot;related sequences&quot; links.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=nuccore&amp;id=21614549,219152114&amp;cmd=ncheck</strong></p><p><strong>cmd=lcheck</strong></p><p>Elink checks for the existence of external links (LinkOuts) for a set of UIDs.</p><p><em>Example: Check whether two protein sequences have any LinkOut providers.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;id=15718680,157427902&amp;cmd=lcheck</strong></p><p><strong>cmd=llinks</strong></p><p>For each input UID, ELink lists the URLs and attributes for the LinkOut providers that are not libraries.</p><p><em>Example: List the LinkOut URLs for non-library providers for two pubmed abstracts.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=19880848,19822630&amp;cmd=llinks</strong></p><p><strong>cmd=llinkslib</strong></p><p>For each input UID, ELink lists the URLs and attributes for <em>all</em> LinkOut providers including libraries.</p><p><em>Example: List all LinkOut URLs for two PubMed abstracts.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=19880848,19822630&amp;cmd=llinkslib</strong></p><p><strong>cmd=prlinks</strong></p><p>ELink lists the primary LinkOut provider for each input UID, or links directly to the LinkOut provider&#39;s web site for a single UID if <strong>retmode</strong> is set to <em>ref</em>.</p><p><em>Example: Find links to full text providers for two PubMed abstracts.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=19880848,19822630&amp;cmd=prlinks</strong></p><p><em>Example: Link directly to the full text for a PubMed abstract at the provider&#39;s web site.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=19880848&amp;cmd=prlinks&amp;retmode=ref</strong></p><h3 id="required-parameter-–-used-only-when-input-is-from-a-uid-list-2" tabindex="-1"><a class="header-anchor" href="#required-parameter-–-used-only-when-input-is-from-a-uid-list-2"><span>Required Parameter – Used only when input is from a UID list</span></a></h3><h4 id="id-3" tabindex="-1"><a class="header-anchor" href="#id-3"><span>id</span></a></h4><p>UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by <strong>dbfrom</strong>. There is no set maximum for the number of UIDs that can be passed to ELink, but if more than about 200 UIDs are to be provided, the request should be made using the HTTP POST method.</p><p><em>Link from protein to gene.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;db=gene&amp;id=15718680,157427902,119703751</strong></p><p><em>Find related sequences (link from nuccore to nuccore).</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=nuccore&amp;db=nuccore&amp;id=34577062</strong></p><p>If more than one <strong>id</strong> parameter is provided, ELink will perform a separate link operation for the set of UIDs specified by each <strong>id</strong> parameter. This effectively accomplishes &quot;one-to-one&quot; links and preserves the connection between the input and output UIDs.</p><p><em>Find one-to-one links from protein to gene.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;db=gene&amp;id=15718680&amp;id=157427902&amp;id=119703751</strong></p><p>For sequence databases (nuccore, popset, protein), the UID list may be a mixed list of GI numbers and accession.version identifiers.</p><p><em>Find one-to-one links from protein to gene.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&amp;db=gene&amp;id=15718680&amp;id=NP_001098858.1&amp;id=119703751</strong></p><h3 id="required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-2" tabindex="-1"><a class="header-anchor" href="#required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-2"><span>Required Parameters – Used only when input is from the Entrez History server</span></a></h3><h4 id="query-key-3" tabindex="-1"><a class="header-anchor" href="#query-key-3"><span>query_key</span></a></h4><p>Query key. This integer specifies which of the UID lists attached to the given Web Environment will be used as input to ELink. Query keys are obtained from the output of previous ESearch, EPost or ELInk calls. The <strong>query_key</strong> parameter must be used in conjunction with <strong>WebEnv</strong>.</p><h4 id="webenv-4" tabindex="-1"><a class="header-anchor" href="#webenv-4"><span>WebEnv</span></a></h4><p>Web Environment. This parameter specifies the Web Environment that contains the UID list to be provided as input to ELink. Usually this WebEnv value is obtained from the output of a previous ESearch, EPost or ELink call. The <strong>WebEnv</strong> parameter must be used in conjunction with <strong>query_key</strong>.</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>Link from protein to gene:</span></span>
<span class="line"><span>elink.fcgi?dbfrom=protein&amp;db=gene&amp;query_key=&lt;key&gt;&amp;WebEnv=&lt;webenv string&gt;</span></span>
<span class="line"><span></span></span>
<span class="line"><span>Find related sequences (link from protein to protein):</span></span>
<span class="line"><span>elink.fcgi?dbfrom=protein&amp;db=protein&amp;query_key=&lt;key&gt;&amp;WebEnv=&lt;webenv string&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="optional-parameter-–-retrieval" tabindex="-1"><a class="header-anchor" href="#optional-parameter-–-retrieval"><span>Optional Parameter – Retrieval</span></a></h3><h4 id="retmode-4" tabindex="-1"><a class="header-anchor" href="#retmode-4"><span>retmode</span></a></h4><p>Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for ELink XML, but ‘json’ is also supported to return output in JSON format.</p><h4 id="idtype-1" tabindex="-1"><a class="header-anchor" href="#idtype-1"><span>idtype</span></a></h4><p>Specifies the type of identifier to return for sequence databases (nuccore, popset, protein). By default, ELink returns GI numbers in its output. If <strong>idtype</strong> is set to ‘acc’, ELink will return accession.version identifiers rather than GI numbers.</p><h3 id="optional-parameters-–-limiting-the-output-set-of-links" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-limiting-the-output-set-of-links"><span>Optional Parameters – Limiting the Output Set of Links</span></a></h3><h4 id="linkname" tabindex="-1"><a class="header-anchor" href="#linkname"><span>linkname</span></a></h4><p>Name of the Entrez link to retrieve. Every link in Entrez is given a name of the form</p><p><em>dbfrom_db_subset</em>.</p><p>The values of <em>subset</em> vary depending on the values of <em>dbfrom</em> and <em>db</em>. Many <em>dbfrom/db</em> combinations have no <em>subset</em> values. See <a href="https://eutils.ncbi.nlm.nih.gov/entrez/query/static/entrezlinks.html" target="_blank" rel="noopener noreferrer">the list of Entrez links</a> for a listing of all available linknames. When <strong>linkname</strong> is used, only the links with that name will be retrieved.</p><p>The <strong>linkname</strong> parameter only functions when <strong>cmd</strong> is set to <em>neighbor</em> or <em>neighbor_history</em>.</p><p><em>Find all links from gene to snp.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=gene&amp;db=snp&amp;id=93986</strong></p><p><em>Find snps with genotype data linked to genes.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=gene&amp;db=snp&amp;id=93986&amp;linkname=gene_snp_genegenotype</strong></p><h4 id="term-1" tabindex="-1"><a class="header-anchor" href="#term-1"><span>term</span></a></h4><p>Entrez query used to limit the output set of linked UIDs. The query in the <strong>term</strong> parameter will be applied after the link operation, and only those UIDs matching the query will be returned by ELink. The <strong>term</strong> parameter only functions when <strong>db</strong> and <strong>dbfrom</strong> are set to the same database value.</p><p><em>Find all related articles for a PMID.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;db=pubmed&amp;id=19879512</strong></p><p><em>Find all related review articles published in 2008 for a PMID.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;db=pubmed&amp;id=19879512&amp;term=review%5Bfilter%5D+AND+2008%5Bpdat%5Dh</strong></p><h4 id="holding" tabindex="-1"><a class="header-anchor" href="#holding"><span>holding</span></a></h4><p>Name of LinkOut provider. Only URLs for the LinkOut provider specified by <strong>holding</strong> will be returned. The value provided to <strong>holding</strong> should be the abbreviation of the LinkOut provider&#39;s name found in the <code>&lt;NameAbbr&gt;</code> tag of the ELink XML output when <strong>cmd</strong> is set to <em>llinks</em> or <em>llinkslib</em>. The <strong>holding</strong> parameter only functions when <strong>cmd</strong> is set to <em>llinks</em> or <em>llinkslib</em>.</p><p><em>Find information for all LinkOut providers for a PMID.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;cmd=llinkslib&amp;id=16210666</strong></p><p><em>Find information from clinicaltrials.gov for a PMID.</em></p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;cmd=llinkslib&amp;id=16210666&amp;holding=CTgov</strong></p><h3 id="optional-parameters-–-dates-1" tabindex="-1"><a class="header-anchor" href="#optional-parameters-–-dates-1"><span>Optional Parameters – Dates</span></a></h3><p>These parameters only function when <strong>cmd</strong> is set to <em>neighbor</em> or <em>neighbor_history</em> and <strong>dbfrom</strong> is <em>pubmed</em>.</p><h4 id="datetype-1" tabindex="-1"><a class="header-anchor" href="#datetype-1"><span>datetype</span></a></h4><p>Type of date used to limit a link operation. The allowed values vary between Entrez databases, but common values are &#39;mdat&#39; (modification date), &#39;pdat&#39; (publication date) and &#39;edat&#39; (Entrez date). Generally an Entrez database will have only two allowed values for <strong>datetype</strong>.</p><h4 id="reldate-1" tabindex="-1"><a class="header-anchor" href="#reldate-1"><span>reldate</span></a></h4><p>When <strong>reldate</strong> is set to an integer <em>n</em>, ELink returns only those items that have a date specified by <strong>datetype</strong> within the last <em>n</em> days.</p><h4 id="mindate-maxdate-1" tabindex="-1"><a class="header-anchor" href="#mindate-maxdate-1"><span>mindate, maxdate</span></a></h4><p>Date range used to limit a link operation by the date specified by <strong>datetype</strong>. These two parameters (<strong>mindate, maxdate</strong>) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM.</p><h2 id="egquery" tabindex="-1"><a class="header-anchor" href="#egquery"><span>EGQuery</span></a></h2><h3 id="base-url-6" tabindex="-1"><a class="header-anchor" href="#base-url-6"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/egquery.fcgi</p><h3 id="function" tabindex="-1"><a class="header-anchor" href="#function"><span>Function</span></a></h3><p>Provides the number of records retrieved in all Entrez databases by a single text query.</p><h3 id="required-parameter-1" tabindex="-1"><a class="header-anchor" href="#required-parameter-1"><span>Required Parameter</span></a></h3><h4 id="term-2" tabindex="-1"><a class="header-anchor" href="#term-2"><span>term</span></a></h4><p>Entrez text query. All special characters must be URL encoded. Spaces may be replaced by &#39;+&#39; signs. For very long queries (more than several hundred characters long), consider using an HTTP POST call. See the <a href="https://www.ncbi.nlm.nih.gov/books/n/helppubmed/pubmedhelp/#pubmedhelp.Search_Field_Descrip" target="_blank" rel="noopener noreferrer">PubMed</a> or <a href="https://www.ncbi.nlm.nih.gov/books/n/helpentrez/EntrezHelp/" target="_blank" rel="noopener noreferrer">Entrez</a> help for information about search field descriptions and tags. Search fields and tags are database specific.</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/egquery.fcgi?term=asthma</strong></p><h2 id="espell" tabindex="-1"><a class="header-anchor" href="#espell"><span>ESpell</span></a></h2><h3 id="base-url-7" tabindex="-1"><a class="header-anchor" href="#base-url-7"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/espell.fcgi</p><h3 id="function-1" tabindex="-1"><a class="header-anchor" href="#function-1"><span>Function</span></a></h3><p>Provides spelling suggestions for terms within a single text query in a given database.</p><h3 id="required-parameters-5" tabindex="-1"><a class="header-anchor" href="#required-parameters-5"><span>Required Parameters</span></a></h3><h4 id="db-6" tabindex="-1"><a class="header-anchor" href="#db-6"><span>db</span></a></h4><p>Database to search. Value must be a valid <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter2/#chapter2.chapter2_table1" target="_blank" rel="noopener noreferrer">Entrez database name</a> (default = pubmed).</p><h4 id="term-3" tabindex="-1"><a class="header-anchor" href="#term-3"><span>term</span></a></h4><p>Entrez text query. All special characters must be URL encoded. Spaces may be replaced by &#39;+&#39; signs. For very long queries (more than several hundred characters long), consider using an HTTP POST call. See the <a href="https://www.ncbi.nlm.nih.gov/books/n/helppubmed/pubmedhelp/#pubmedhelp.Search_Field_Descrip" target="_blank" rel="noopener noreferrer">PubMed</a> or <a href="https://www.ncbi.nlm.nih.gov/books/n/helpentrez/EntrezHelp/" target="_blank" rel="noopener noreferrer">Entrez</a> help for information about search field descriptions and tags. Search fields and tags are database specific.</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/espell.fcgi?db=pubmed&amp;term=asthmaa+OR+alergies</strong></p><h2 id="ecitmatch" tabindex="-1"><a class="header-anchor" href="#ecitmatch"><span>ECitMatch</span></a></h2><h3 id="base-url-8" tabindex="-1"><a class="header-anchor" href="#base-url-8"><span>Base URL</span></a></h3><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ecitmatch.cgi</p><h3 id="function-2" tabindex="-1"><a class="header-anchor" href="#function-2"><span>Function</span></a></h3><p>Retrieves PubMed IDs (PMIDs) that correspond to a set of input citation strings.</p><h3 id="required-parameters-6" tabindex="-1"><a class="header-anchor" href="#required-parameters-6"><span>Required Parameters</span></a></h3><h4 id="db-7" tabindex="-1"><a class="header-anchor" href="#db-7"><span>db</span></a></h4><p>Database to search. The only supported value is ‘pubmed’.</p><h4 id="rettype-2" tabindex="-1"><a class="header-anchor" href="#rettype-2"><span>rettype</span></a></h4><p>Retrieval type. The only supported value is ‘xml’.</p><h4 id="bdata" tabindex="-1"><a class="header-anchor" href="#bdata"><span>bdata</span></a></h4><p>Citation strings. Each input citation must be represented by a citation string in the following format:</p><p>journal_title|year|volume|first_page|author_name|your_key|</p><p>Multiple citation strings may be provided by separating the strings with a carriage return character (%0D). The <em>your_key</em> value is an arbitrary label provided by the user that may serve as a local identifier for the citation, and it will be included in the output. Be aware that all spaces must be replaced by ‘+’ symbols and that citation strings should end with a final vertical bar ‘|’.</p><p><strong>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ecitmatch.cgi?db=pubmed&amp;retmode=xml&amp;bdata=proc+natl+acad+sci+u+s+a|1991|88|3248|mann+bj|Art1|%0Dscience|1987|235|182|palmenberg+ac|Art2|</strong></p><h2 id="release-notes" tabindex="-1"><a class="header-anchor" href="#release-notes"><span>Release Notes</span></a></h2><h3 id="efetch-elink-json-ouput-june-24-2015" tabindex="-1"><a class="header-anchor" href="#efetch-elink-json-ouput-june-24-2015"><span>EFetch; ELink JSON ouput: June 24, 2015</span></a></h3><ul><li>EFetch now supports ClinVar and GTR</li><li>ELink now provides output in JSON format</li></ul><h3 id="esearch-sort-json-output-format-february-14-2014" tabindex="-1"><a class="header-anchor" href="#esearch-sort-json-output-format-february-14-2014"><span>ESearch &amp;sort; JSON output format: February 14, 2014</span></a></h3><ul><li>ESearch now provides a supported <strong>sort</strong> parameter</li><li>EInfo, ESearch and ESummary now provide output data in JSON format</li></ul><h3 id="ecitmatch-einfo-version-2-0-efetch-august-9-2013" tabindex="-1"><a class="header-anchor" href="#ecitmatch-einfo-version-2-0-efetch-august-9-2013"><span>ECitMatch, EInfo Version 2.0, EFetch: August 9, 2013</span></a></h3><ul><li>ECitMatch is a new E-utility that serves as an API to the PubMed <a href="https://www.ncbi.nlm.nih.gov/pubmed/batchcitmatch" target="_blank" rel="noopener noreferrer">batch citation matcher</a></li><li>EInfo has an updated XML output that includes two new fields: <code>&lt;IsTruncatable&gt;</code> and <code>&lt;IsRangeable&gt;</code></li><li>EFetch now supports the BioProject database.</li></ul><h3 id="efetch-version-2-0-target-release-date-february-15-2012" tabindex="-1"><a class="header-anchor" href="#efetch-version-2-0-target-release-date-february-15-2012"><span>EFetch Version 2.0. Target release date: February 15, 2012</span></a></h3><ul><li>EFetch now supports the following databases: biosample, biosystems and sra</li><li>EFetch now has defined default values for &amp;retmode and &amp;rettype for all supported databases (please see <a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly" target="_blank" rel="noopener noreferrer">Table 1</a> for all supported values of these parameters)</li><li>EFetch no longer supports &amp;retmode=html; requests containing &amp;retmode=html will return data using the default &amp;retmode value for the specified database (&amp;db)</li><li>EFetch requests including &amp;rettype=docsum return XML data equivalent to ESummary output</li></ul><h3 id="release-of-new-genome-database-november-9-2011" tabindex="-1"><a class="header-anchor" href="#release-of-new-genome-database-november-9-2011"><span>Release of new Genome database: November 9, 2011</span></a></h3><ul><li>Entrez Genome has been completely redesigned, and database records now correspond to a species rather than an individual chromosome sequence. Please see full details of the change at https://www.ncbi.nlm.nih.gov/About/news/17Nov2011.html</li><li>Old Genome IDs are no longer valid. A file is available on the NCBI FTP site that maps old Genome IDs to Nucleotide GIs: <a href="https://ftp.ncbi.nih.gov/genomes/old_genomeID2nucGI" target="_blank" rel="noopener noreferrer">ftp.ncbi.nih.gov/genomes/old_genomeID2nucGI</a></li><li>EFetch no longer supports retrievals from Genome (db=genome).</li><li>The ESummary XML for Genome has been recast to reflect the new data model.</li><li>To view the new search fields and links supported for the new Genome database, please see https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=genome</li></ul><h3 id="esummary-version-2-0-november-4-2011" tabindex="-1"><a class="header-anchor" href="#esummary-version-2-0-november-4-2011"><span>ESummary Version 2.0. November 4, 2011</span></a></h3><ul><li>ESummary now supports a new, alternative XML presentation for Entrez document summaries (DocSums). The new XML is unique to each Entrez database and generally contains more extensive data about the record than the original DocSum XML.</li><li>There are no plans at present to discontinue the original DocSum XML, so developers can continue to use this presentation, which will remain the default.</li><li>Version 2.0 XML is returned when &amp;version=2.0 is included in the ESummary URL.</li></ul><h2 id="demonstration-programs" tabindex="-1"><a class="header-anchor" href="#demonstration-programs"><span>Demonstration Programs</span></a></h2><p>Please see <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter1/#chapter1.Demonstration_Programs" target="_blank" rel="noopener noreferrer">Chapter 1</a> for sample Perl scripts.</p><h2 id="for-more-information" tabindex="-1"><a class="header-anchor" href="#for-more-information"><span>For More Information</span></a></h2><p>Please see <a href="https://www.ncbi.nlm.nih.gov/books/n/helpeutils/chapter1/#chapter1.For_More_Information_8" target="_blank" rel="noopener noreferrer">Chapter 1</a> for getting additional information about the E-utilities.</p>`,403),i=[s];function o(l,h){return a(),t("div",null,i)}const c=e(n,[["render",o],["__file","04-The E-utilities In-Depth_EN.html.vue"]]),u=JSON.parse('{"path":"/E-utilities_EN/04-The%20E-utilities%20In-Depth_EN.html","title":"The E-utilities In-Depth: Parameters, Syntax and More","lang":"zh-CN","frontmatter":{"order":4,"description":"The E-utilities In-Depth: Parameters, Syntax and More Eric Sayers, PhD. Author Information and Affiliations Authors Eric Sayers, PhDcorresponding author1. Affilitions 1 NCBI Ema...","head":[["meta",{"property":"og:url","content":"https://vuepress-theme-hope-docs-demo.netlify.app/E-utilities_zh-CN/E-utilities_EN/04-The%20E-utilities%20In-Depth_EN.html"}],["meta",{"property":"og:site_name","content":"E-utilities 汉化文档"}],["meta",{"property":"og:title","content":"The E-utilities In-Depth: Parameters, Syntax and More"}],["meta",{"property":"og:description","content":"The E-utilities In-Depth: Parameters, Syntax and More Eric Sayers, PhD. Author Information and Affiliations Authors Eric Sayers, PhDcorresponding author1. Affilitions 1 NCBI Ema..."}],["meta",{"property":"og:type","content":"article"}],["meta",{"property":"og:image","content":"https://www.ncbi.nlm.nih.gov/corehtml/pmc/pmcgifs/corrauth.gif"}],["meta",{"property":"og:locale","content":"zh-CN"}],["meta",{"property":"og:updated_time","content":"2024-06-21T12:28:18.000Z"}],["meta",{"property":"article:author","content":"WhyLIM"}],["meta",{"property":"article:modified_time","content":"2024-06-21T12:28:18.000Z"}],["script",{"type":"application/ld+json"},"{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Article\\",\\"headline\\":\\"The E-utilities In-Depth: Parameters, Syntax and More\\",\\"image\\":[\\"https://www.ncbi.nlm.nih.gov/corehtml/pmc/pmcgifs/corrauth.gif\\",\\"https://www.ncbi.nlm.nih.gov/corehtml/pmc/pmcgifs/corrauth.gif\\",\\"https://www.ncbi.nlm.nih.gov/corehtml/pmc/css/bookshelf/2.26/img/table-icon.gif\\"],\\"dateModified\\":\\"2024-06-21T12:28:18.000Z\\",\\"author\\":[{\\"@type\\":\\"Person\\",\\"name\\":\\"WhyLIM\\",\\"url\\":\\"https://github.com/WhyLIM\\"}]}"]]},"headers":[{"level":2,"title":"Introduction","slug":"introduction","link":"#introduction","children":[]},{"level":2,"title":"General Usage Guidelines","slug":"general-usage-guidelines","link":"#general-usage-guidelines","children":[{"level":3,"title":"tool","slug":"tool","link":"#tool","children":[]},{"level":3,"title":"email","slug":"email","link":"#email","children":[]},{"level":3,"title":"api_key","slug":"api-key","link":"#api-key","children":[]}]},{"level":2,"title":"E-utilities DTDs","slug":"e-utilities-dtds","link":"#e-utilities-dtds","children":[]},{"level":2,"title":"EInfo","slug":"einfo","link":"#einfo","children":[{"level":3,"title":"Base URL","slug":"base-url","link":"#base-url","children":[]},{"level":3,"title":"Functions","slug":"functions","link":"#functions","children":[]},{"level":3,"title":"Required Parameters","slug":"required-parameters","link":"#required-parameters","children":[]},{"level":3,"title":"Optional Parameters","slug":"optional-parameters","link":"#optional-parameters","children":[]},{"level":3,"title":"Examples","slug":"examples","link":"#examples","children":[]}]},{"level":2,"title":"ESearch","slug":"esearch","link":"#esearch","children":[{"level":3,"title":"Base URL","slug":"base-url-1","link":"#base-url-1","children":[]},{"level":3,"title":"Functions","slug":"functions-1","link":"#functions-1","children":[]},{"level":3,"title":"Required Parameters","slug":"required-parameters-1","link":"#required-parameters-1","children":[]},{"level":3,"title":"Optional Parameters – History Server","slug":"optional-parameters-–-history-server","link":"#optional-parameters-–-history-server","children":[]},{"level":3,"title":"Optional Parameters – Retrieval","slug":"optional-parameters-–-retrieval","link":"#optional-parameters-–-retrieval","children":[]},{"level":3,"title":"Optional Parameters – Dates","slug":"optional-parameters-–-dates","link":"#optional-parameters-–-dates","children":[]},{"level":3,"title":"Examples","slug":"examples-1","link":"#examples-1","children":[]}]},{"level":2,"title":"EPost","slug":"epost","link":"#epost","children":[{"level":3,"title":"Base URL","slug":"base-url-2","link":"#base-url-2","children":[]},{"level":3,"title":"Functions","slug":"functions-2","link":"#functions-2","children":[]},{"level":3,"title":"Required Parameters","slug":"required-parameters-2","link":"#required-parameters-2","children":[]},{"level":3,"title":"Optional Parameter","slug":"optional-parameter","link":"#optional-parameter","children":[]},{"level":3,"title":"Example","slug":"example","link":"#example","children":[]}]},{"level":2,"title":"ESummary","slug":"esummary","link":"#esummary","children":[{"level":3,"title":"Base URL","slug":"base-url-3","link":"#base-url-3","children":[]},{"level":3,"title":"Functions","slug":"functions-3","link":"#functions-3","children":[]},{"level":3,"title":"Required Parameter","slug":"required-parameter","link":"#required-parameter","children":[]},{"level":3,"title":"Required Parameter – Used only when input is from a UID list","slug":"required-parameter-–-used-only-when-input-is-from-a-uid-list","link":"#required-parameter-–-used-only-when-input-is-from-a-uid-list","children":[]},{"level":3,"title":"Required Parameters – Used only when input is from the Entrez History server","slug":"required-parameters-–-used-only-when-input-is-from-the-entrez-history-server","link":"#required-parameters-–-used-only-when-input-is-from-the-entrez-history-server","children":[]},{"level":3,"title":"Optional Parameters – Retrieval","slug":"optional-parameters-–-retrieval-1","link":"#optional-parameters-–-retrieval-1","children":[]},{"level":3,"title":"Examples","slug":"examples-2","link":"#examples-2","children":[]}]},{"level":2,"title":"EFetch","slug":"efetch","link":"#efetch","children":[{"level":3,"title":"Base URL","slug":"base-url-4","link":"#base-url-4","children":[]},{"level":3,"title":"Functions","slug":"functions-4","link":"#functions-4","children":[]},{"level":3,"title":"Required Parameters","slug":"required-parameters-3","link":"#required-parameters-3","children":[]},{"level":3,"title":"Required Parameter – Used only when input is from a UID list","slug":"required-parameter-–-used-only-when-input-is-from-a-uid-list-1","link":"#required-parameter-–-used-only-when-input-is-from-a-uid-list-1","children":[]},{"level":3,"title":"Required Parameters – Used only when input is from the Entrez History server","slug":"required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-1","link":"#required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-1","children":[]},{"level":3,"title":"Optional Parameters – Retrieval","slug":"optional-parameters-–-retrieval-2","link":"#optional-parameters-–-retrieval-2","children":[]},{"level":3,"title":"Optional Parameters – Sequence Databases","slug":"optional-parameters-–-sequence-databases","link":"#optional-parameters-–-sequence-databases","children":[]},{"level":3,"title":"Examples","slug":"examples-3","link":"#examples-3","children":[]}]},{"level":2,"title":"ELink","slug":"elink","link":"#elink","children":[{"level":3,"title":"Base URL","slug":"base-url-5","link":"#base-url-5","children":[]},{"level":3,"title":"Functions","slug":"functions-5","link":"#functions-5","children":[]},{"level":3,"title":"Required Parameters","slug":"required-parameters-4","link":"#required-parameters-4","children":[]},{"level":3,"title":"Required Parameter – Used only when input is from a UID list","slug":"required-parameter-–-used-only-when-input-is-from-a-uid-list-2","link":"#required-parameter-–-used-only-when-input-is-from-a-uid-list-2","children":[]},{"level":3,"title":"Required Parameters – Used only when input is from the Entrez History server","slug":"required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-2","link":"#required-parameters-–-used-only-when-input-is-from-the-entrez-history-server-2","children":[]},{"level":3,"title":"Optional Parameter – Retrieval","slug":"optional-parameter-–-retrieval","link":"#optional-parameter-–-retrieval","children":[]},{"level":3,"title":"Optional Parameters – Limiting the Output Set of Links","slug":"optional-parameters-–-limiting-the-output-set-of-links","link":"#optional-parameters-–-limiting-the-output-set-of-links","children":[]},{"level":3,"title":"Optional Parameters – Dates","slug":"optional-parameters-–-dates-1","link":"#optional-parameters-–-dates-1","children":[]}]},{"level":2,"title":"EGQuery","slug":"egquery","link":"#egquery","children":[{"level":3,"title":"Base URL","slug":"base-url-6","link":"#base-url-6","children":[]},{"level":3,"title":"Function","slug":"function","link":"#function","children":[]},{"level":3,"title":"Required Parameter","slug":"required-parameter-1","link":"#required-parameter-1","children":[]}]},{"level":2,"title":"ESpell","slug":"espell","link":"#espell","children":[{"level":3,"title":"Base URL","slug":"base-url-7","link":"#base-url-7","children":[]},{"level":3,"title":"Function","slug":"function-1","link":"#function-1","children":[]},{"level":3,"title":"Required Parameters","slug":"required-parameters-5","link":"#required-parameters-5","children":[]}]},{"level":2,"title":"ECitMatch","slug":"ecitmatch","link":"#ecitmatch","children":[{"level":3,"title":"Base URL","slug":"base-url-8","link":"#base-url-8","children":[]},{"level":3,"title":"Function","slug":"function-2","link":"#function-2","children":[]},{"level":3,"title":"Required Parameters","slug":"required-parameters-6","link":"#required-parameters-6","children":[]}]},{"level":2,"title":"Release Notes","slug":"release-notes","link":"#release-notes","children":[{"level":3,"title":"EFetch; ELink JSON ouput: June 24, 2015","slug":"efetch-elink-json-ouput-june-24-2015","link":"#efetch-elink-json-ouput-june-24-2015","children":[]},{"level":3,"title":"ESearch &sort; JSON output format: February 14, 2014","slug":"esearch-sort-json-output-format-february-14-2014","link":"#esearch-sort-json-output-format-february-14-2014","children":[]},{"level":3,"title":"ECitMatch, EInfo Version 2.0, EFetch: August 9, 2013","slug":"ecitmatch-einfo-version-2-0-efetch-august-9-2013","link":"#ecitmatch-einfo-version-2-0-efetch-august-9-2013","children":[]},{"level":3,"title":"EFetch Version 2.0. Target release date: February 15, 2012","slug":"efetch-version-2-0-target-release-date-february-15-2012","link":"#efetch-version-2-0-target-release-date-february-15-2012","children":[]},{"level":3,"title":"Release of new Genome database: November 9, 2011","slug":"release-of-new-genome-database-november-9-2011","link":"#release-of-new-genome-database-november-9-2011","children":[]},{"level":3,"title":"ESummary Version 2.0. November 4, 2011","slug":"esummary-version-2-0-november-4-2011","link":"#esummary-version-2-0-november-4-2011","children":[]}]},{"level":2,"title":"Demonstration Programs","slug":"demonstration-programs","link":"#demonstration-programs","children":[]},{"level":2,"title":"For More Information","slug":"for-more-information","link":"#for-more-information","children":[]}],"git":{"createdTime":1718972898000,"updatedTime":1718972898000,"contributors":[{"name":"Mli-TB","email":"mli.bio@outlook.com","commits":1}]},"readingTime":{"minutes":20.14,"words":6042},"filePathRelative":"E-utilities_EN/04-The E-utilities In-Depth_EN.md","localizedDate":"2024年6月21日","autoDesc":true}');export{c as comp,u as data};
