import{_ as s}from"./plugin-vue_export-helper-DlAUqK2U.js";import{c as n,o as a,d as e}from"./app-BSJRkk3t.js";const i={},l=e(`<h1 id="entrez-direct-examples" tabindex="-1"><a class="header-anchor" href="#entrez-direct-examples"><span>Entrez Direct Examples</span></a></h1><blockquote><p>Jonathan Kans, PhD.</p><p><a href="https://www.ncbi.nlm.nih.gov/books/NBK565821/#__NBK565821_ai__" target="_blank" rel="noopener noreferrer">Author Information and Affiliations</a></p><p>Created: April 23, 2013; Last Update: April 12, 2024.</p><p><em>Estimated reading time: 21 minutes</em></p></blockquote><p>Some early EDirect examples required post-processing with shell script:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  start=$(( start + 1 ))</span></span>
<span class="line"><span>  stop=$(( stop + 1 ))</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>or awk commands:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  awk -F &#39;\\t&#39; -v &#39;OFS=\\t&#39; &#39;{print $1, $2+1, $3+1}&#39;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>to increment 0-based sequence coordinates so that the proper region was retrieved:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id $accn -format fasta -seq_start $start -seq_stop $stop</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>This led to -element derivatives that could do simple adjustments within an xtract command:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  -element ChrAccVer -inc ChrStart -1-based ChrStop</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>and to efetch arguments that could take 0-based coordinates directly:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id $accn -format fasta -chr_start $start -chr_stop $stop</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>These helped eliminate the need for scripts to perform otherwise trivial modifications on extracted data.</p><p>More recent work allows easier after-the-fact numeric manipulation using the filter-columns:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  filter-columns &#39;10 &lt;= $2 &amp;&amp; $2 &lt;= 30&#39;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>and print-columns:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  print-columns &#39;$1, $2+1, $3-1, &quot;\\042&quot; $4 &quot;\\042&quot;, tolower($5), log($3), total += $2&#39;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>scripts, which accept column designators in an argument and pass them to internal awk commands. The NF (number of fields) and NR (current record number) built-in variables can also be used, as can YR (for year) and DT (for date in YYYY-MM-DD format).</p><h2 id="pubmed" tabindex="-1"><a class="header-anchor" href="#pubmed"><span>PubMed</span></a></h2><h3 id="author-frequency" tabindex="-1"><a class="header-anchor" href="#author-frequency"><span>Author Frequency</span></a></h3><p>Who are the most prolific authors on rattlesnake phospholipase?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db pubmed -query \\</span></span>
<span class="line"><span>    &quot;crotalid venoms [MAJR] AND phospholipase [TIAB]&quot; |</span></span>
<span class="line"><span>  efetch -format xml |</span></span>
<span class="line"><span>  xtract -pattern PubmedArticle \\</span></span>
<span class="line"><span>    -block Author -sep &quot; &quot; -tab &quot;\\n&quot; -element LastName,Initials |</span></span>
<span class="line"><span>  sort-uniq-count-rank</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Placing author names on separate lines allows sort-uniq-count-rank to produce a frequency table:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  87    Lomonte B</span></span>
<span class="line"><span>  77    Gutiérrez JM</span></span>
<span class="line"><span>  64    Soares AM</span></span>
<span class="line"><span>  53    Marangoni S</span></span>
<span class="line"><span>  43    Giglio JR</span></span>
<span class="line"><span>  39    Bon C</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="publication-distribution" tabindex="-1"><a class="header-anchor" href="#publication-distribution"><span>Publication Distribution</span></a></h3><p>When were the most papers about Legionnaires disease published?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db pubmed -query &quot;legionnaires disease [TITL]&quot; |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element PubDate |</span></span>
<span class="line"><span>  cut -c 1-4 |</span></span>
<span class="line"><span>  sort-uniq-count-rank</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>In this case sort-uniq-count-rank reports the number of selected papers per year:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  173    1979</span></span>
<span class="line"><span>  102    1980</span></span>
<span class="line"><span>  96     1978</span></span>
<span class="line"><span>  92     1981</span></span>
<span class="line"><span>  66     1983</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Using ‑year &quot;PubDate/*&quot; on a PubmedArticle record takes the first four-digit number it encounters, and the result does not need to be trimmed:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db pubmed -query &quot;legionnaires disease [TITL]&quot; |</span></span>
<span class="line"><span>  efetch -format xml |</span></span>
<span class="line"><span>  xtract -pattern PubmedArticle -year &quot;PubDate/*&quot; |</span></span>
<span class="line"><span>  sort-uniq-count-rank</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="treatment-locations" tabindex="-1"><a class="header-anchor" href="#treatment-locations"><span>Treatment Locations</span></a></h3><p>What is the geographic distribution of sepsis treatment studies?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db pubmed -query \\</span></span>
<span class="line"><span>    &quot;sepsis/therapy [MESH] AND geographic locations [MESH]&quot; |</span></span>
<span class="line"><span>  efetch -format xml |</span></span>
<span class="line"><span>  xtract -pattern PubmedArticle \\</span></span>
<span class="line"><span>    -block MeshHeading -if DescriptorName@Type -equals Geographic \\</span></span>
<span class="line"><span>      -tab &quot;\\n&quot; -element DescriptorName |</span></span>
<span class="line"><span>  sort-uniq-count-rank</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This returns the number of articles ranked by country or region:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  660    United States</span></span>
<span class="line"><span>  250    Spain</span></span>
<span class="line"><span>  182    Germany</span></span>
<span class="line"><span>  168    India</span></span>
<span class="line"><span>  155    Taiwan</span></span>
<span class="line"><span>  139    Japan</span></span>
<span class="line"><span>  126    China</span></span>
<span class="line"><span>  124    France</span></span>
<span class="line"><span>  123    Europe</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Note that England and United Kingdom will appear as two separate entries.</p><h3 id="indexed-date-fields" tabindex="-1"><a class="header-anchor" href="#indexed-date-fields"><span>Indexed Date Fields</span></a></h3><p>What date fields are indexed for PubMed?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  einfo -db pubmed |</span></span>
<span class="line"><span>  xtract -pattern Field \\</span></span>
<span class="line"><span>    -if IsDate -equals Y -and IsHidden -equals N \\</span></span>
<span class="line"><span>      -pfx &quot;[&quot; -sfx &quot;]&quot; -element Name \\</span></span>
<span class="line"><span>      -pfx &quot;&quot; -sfx &quot;&quot; -element FullName |</span></span>
<span class="line"><span>  sort -k 2f | expand</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Indexed dates are shown with field abbreviations and descriptive names:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  [CDAT]  Date - Completion</span></span>
<span class="line"><span>  [CRDT]  Date - Create</span></span>
<span class="line"><span>  [EDAT]  Date - Entrez</span></span>
<span class="line"><span>  [MHDA]  Date - MeSH</span></span>
<span class="line"><span>  [MDAT]  Date - Modification</span></span>
<span class="line"><span>  [PDAT]  Date - Publication</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="digital-object-identifiers" tabindex="-1"><a class="header-anchor" href="#digital-object-identifiers"><span>Digital Object Identifiers</span></a></h3><p>How are digital object identifiers obtained from PubMed articles?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db pubmed -query &quot;Rowley JD [AUTH]&quot; |</span></span>
<span class="line"><span>  efetch -format xml |</span></span>
<span class="line"><span>  xtract -head &#39;&lt;html&gt;&lt;body&gt;&#39; -tail &#39;&lt;/body&gt;&lt;/html&gt;&#39; \\</span></span>
<span class="line"><span>    -pattern PubmedArticle -PMID MedlineCitation/PMID \\</span></span>
<span class="line"><span>      -block ArticleId -if @IdType -equals doi \\</span></span>
<span class="line"><span>        -tab &quot;&quot; -pfx &#39;&lt;p&gt;&lt;a href=&quot;&#39; -sfx &#39;&quot;&gt;&#39; -doi ArticleId \\</span></span>
<span class="line"><span>        -tab &quot;\\n&quot; -pfx &#39;&#39; -sfx &#39;&lt;/a&gt;&lt;/p&gt;&#39; -element &quot;&amp;PMID&quot; |</span></span>
<span class="line"><span>  transmute -format</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>The ‑doi command extracts the DOIs and constructs URL references. The ‑pfx and ‑sfx arguments used here enclose each PMID with a clickable link to its DOI:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot; ?&gt;</span></span>
<span class="line"><span>  &lt;!DOCTYPE html&gt;</span></span>
<span class="line"><span>  &lt;html&gt;</span></span>
<span class="line"><span>    &lt;body&gt;</span></span>
<span class="line"><span>      &lt;p&gt;</span></span>
<span class="line"><span>        &lt;a href=&quot;https://doi.org/10.1038%2Fleu.2013.340&quot;&gt;24496283&lt;/a&gt;</span></span>
<span class="line"><span>      &lt;/p&gt;</span></span>
<span class="line"><span>      &lt;p&gt;</span></span>
<span class="line"><span>        &lt;a href=&quot;https://doi.org/10.1073%2Fpnas.1310656110&quot;&gt;23818607&lt;/a&gt;</span></span>
<span class="line"><span>      &lt;/p&gt;</span></span>
<span class="line"><span>      ...</span></span>
<span class="line"><span>    &lt;/body&gt;</span></span>
<span class="line"><span>  &lt;/html&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="reference-formatting" tabindex="-1"><a class="header-anchor" href="#reference-formatting"><span>Reference Formatting</span></a></h3><p>How were references in the main EDirect documentation page formatted?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db pubmed -format docsum -id 26931183 31738401 31600197 \\</span></span>
<span class="line"><span>    29718389 14728215 11449725 8743683 31114887 23175613 |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>Relevant document summary fields were collected with:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -set Set -rec Rec -pattern DocumentSummary \\</span></span>
<span class="line"><span>    -sep &quot;, &quot; -sfx &quot;.&quot; -NM Name -clr \\</span></span>
<span class="line"><span>    -sfx &quot;.&quot; -SRC Source -clr \\</span></span>
<span class="line"><span>    -pfx &quot;:&quot; -PG Pages -clr \\</span></span>
<span class="line"><span>    -pfx &quot;(PMID &quot; -sfx &quot;.)&quot; -ID Id -clr \\</span></span>
<span class="line"><span>    -group ArticleId -if IdType -equals doi \\</span></span>
<span class="line"><span>      -pfx &quot;https://doi.org/&quot; -sfx &quot;.&quot; -DOI Value -PG &quot;(.)&quot; -clr \\</span></span>
<span class="line"><span>    -group DocumentSummary \\</span></span>
<span class="line"><span>      -wrp Sort -lower &quot;&amp;NM&quot; \\</span></span>
<span class="line"><span>      -wrp Name -element &quot;&amp;NM&quot; \\</span></span>
<span class="line"><span>      -wrp Title -element Title \\</span></span>
<span class="line"><span>      -wrp Source -element &quot;&amp;SRC&quot; \\</span></span>
<span class="line"><span>      -wrp Year -year PubDate \\</span></span>
<span class="line"><span>      -wrp Pages -element &quot;&amp;PG&quot; \\</span></span>
<span class="line"><span>      -wrp DOI -element &quot;&amp;DOI&quot; \\</span></span>
<span class="line"><span>      -wrp Id -element &quot;&amp;ID&quot; |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This generated intermediate XML with partial formatting:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  ...</span></span>
<span class="line"><span>  &lt;Rec&gt;</span></span>
<span class="line"><span>    &lt;Sort&gt;wei ch, allot a, leaman r, lu z.&lt;/Sort&gt;</span></span>
<span class="line"><span>    &lt;Name&gt;Wei CH, Allot A, Leaman R, Lu Z.&lt;/Name&gt;</span></span>
<span class="line"><span>    &lt;Title&gt;PubTator central: automated concept ... full text articles.&lt;/Title&gt;</span></span>
<span class="line"><span>    &lt;Source&gt;Nucleic Acids Res.&lt;/Source&gt;</span></span>
<span class="line"><span>    &lt;Year&gt;2019&lt;/Year&gt;</span></span>
<span class="line"><span>    &lt;Pages&gt;.&lt;/Pages&gt;</span></span>
<span class="line"><span>    &lt;DOI&gt;https://doi.org/10.1093/nar/gkz389.&lt;/DOI&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;(PMID 31114887.)&lt;/Id&gt;</span></span>
<span class="line"><span>  &lt;/Rec&gt;</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>References were sorted by the first author’s last name:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -set Set -rec Rec -pattern Rec -sort Sort |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>and then printed with:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -pattern Rec -deq &quot;\\n\\n&quot; -tab &quot; &quot; -sep &quot;&quot; \\</span></span>
<span class="line"><span>    -element Name Title Source Year,Pages DOI Id</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>followed by manual editing of hyphenated initials:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  ...</span></span>
<span class="line"><span>  Wei C-H, Allot A, Leaman R, Lu Z. PubTator central: automated concept</span></span>
<span class="line"><span>  annotation for biomedical full text articles. Nucleic Acids Res. 2019.</span></span>
<span class="line"><span>  https://doi.org/10.1093/nar/gkz389. (PMID 31114887.)</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Using xtract <strong>‑reg</strong> and <strong>‑exp</strong> regular expressions to guide xtract <strong>‑replace</strong>:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db pubmed -id 31114887 -format xml |</span></span>
<span class="line"><span>  xtract -pattern PubmedArticle -block Author -deq &quot;\\n&quot; \\</span></span>
<span class="line"><span>    -element LastName -reg &quot;[a-z .]&quot; -exp &quot;&quot; -replace ForeName</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>can generate hyphenated initials directly from the PubmedArticle ForeName field:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &lt;Author&gt;</span></span>
<span class="line"><span>    &lt;LastName&gt;Wei&lt;/LastName&gt;</span></span>
<span class="line"><span>    &lt;ForeName&gt;Chih-Hsuan&lt;/ForeName&gt;</span></span>
<span class="line"><span>    &lt;Initials&gt;CH&lt;/Initials&gt;</span></span>
<span class="line"><span>  &lt;/Author&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>by removing lower-case letters, spaces, and periods.</p><h2 id="nucleotide" tabindex="-1"><a class="header-anchor" href="#nucleotide"><span>Nucleotide</span></a></h2><h3 id="coding-sequences" tabindex="-1"><a class="header-anchor" href="#coding-sequences"><span>Coding Sequences</span></a></h3><p>What are the coding sequences in the <em>Escherichia coli</em> lac operon?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id J01636.1 -format gbc |</span></span>
<span class="line"><span>  xtract -insd CDS gene sub_sequence</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>Sequence under a feature location is obtained with the ‑insd &quot;sub_sequence&quot; argument:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  J01636.1    lacI    GTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCG...</span></span>
<span class="line"><span>  J01636.1    lacZ    ATGACCATGATTACGGATTCACTGGCCGTCGTTTTACAAC...</span></span>
<span class="line"><span>  J01636.1    lacY    ATGTACTATTTAAAAAACACAAACTTTTGGATGTTCGGTT...</span></span>
<span class="line"><span>  J01636.1    lacA    TTGAACATGCCAATGACCGAAAGAATAAGAGCAGGCAAGC...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="untranslated-region-sequences" tabindex="-1"><a class="header-anchor" href="#untranslated-region-sequences"><span>Untranslated Region Sequences</span></a></h3><p>What are the 5&#39; and 3&#39; UTR sequences for lycopene cyclase mRNAs?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  SubSeq() {</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    xtract -pattern INSDSeq -ACC INSDSeq_accession-version -SEQ INSDSeq_sequence \\</span></span>
<span class="line"><span>      -group INSDFeature -if INSDFeature_key -equals CDS -PRD &quot;(-)&quot; \\</span></span>
<span class="line"><span>        -block INSDQualifier -if INSDQualifier_name -equals product \\</span></span>
<span class="line"><span>          -PRD INSDQualifier_value \\</span></span>
<span class="line"><span>        -block INSDFeature -pfc &quot;\\n&quot; -element &quot;&amp;ACC&quot; -rst \\</span></span>
<span class="line"><span>          -first INSDInterval_from -last INSDInterval_to -element &quot;&amp;PRD&quot; &quot;&amp;SEQ&quot;</span></span>
<span class="line"><span>  }</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  UTRs() {</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    while IFS=$&#39;\\t&#39; read acc fst lst prd seq</span></span>
<span class="line"><span>    do</span></span>
<span class="line"><span>      if [ $fst -gt 1 ]</span></span>
<span class="line"><span>      then</span></span>
<span class="line"><span>        echo -e &quot;&gt;$acc 5&#39;UTR: 1..$((fst-1)) $prd&quot;</span></span>
<span class="line"><span>        echo &quot;\${seq:1:$((fst-2))}&quot; | fold -w 50</span></span>
<span class="line"><span>      else</span></span>
<span class="line"><span>        echo -e &quot;&gt;$acc NO 5&#39;UTR&quot;</span></span>
<span class="line"><span>      fi</span></span>
<span class="line"><span>      if [ $lst -lt \${#seq} ]</span></span>
<span class="line"><span>      then</span></span>
<span class="line"><span>        echo -e &quot;&gt;$acc 3&#39;UTR: $((lst+1))..\${#seq} $prd&quot;</span></span>
<span class="line"><span>        echo &quot;\${seq:$lst}&quot; | fold -w 50</span></span>
<span class="line"><span>      else</span></span>
<span class="line"><span>        echo -e &quot;&gt;$acc NO 3&#39;UTR&quot;</span></span>
<span class="line"><span>      fi</span></span>
<span class="line"><span>    done</span></span>
<span class="line"><span>  }</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  esearch -db nuccore -query &quot;5.5.1.19 [ECNO]&quot; |</span></span>
<span class="line"><span>  efilter -molecule mrna -feature cds -source refseq |</span></span>
<span class="line"><span>  efetch -format gbc |</span></span>
<span class="line"><span>  SubSeq | UTRs</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Sequences before the start codon and after the stop codon are obtained with Unix substring commands:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &gt;NM_001324787.1 NO 5&#39;UTR</span></span>
<span class="line"><span>  &gt;NM_001324787.1 NO 3&#39;UTR</span></span>
<span class="line"><span>  &gt;NM_001324979.1 5&#39;UTR: 1..262 lycopene beta cyclase, chloroplastic/chromoplastic</span></span>
<span class="line"><span>  attatagaaatacttaagatatatcattgccctttaatcatttattttta</span></span>
<span class="line"><span>  actcttttaagtgtttaaagattgattctttgtacatgttctgcttcatt</span></span>
<span class="line"><span>  tgtgttgaaaattgagttgttttcttgaattttgcaagaatataggggac</span></span>
<span class="line"><span>  cccatttgtgttgaaaattgagcagctttctttgtgttttgttcgatttt</span></span>
<span class="line"><span>  tcaagaatataggaccccattttctgttttcttgagataaattgcacctt</span></span>
<span class="line"><span>  gttgggaaaat</span></span>
<span class="line"><span>  &gt;NM_001324979.1 3&#39;UTR: 1760..1782 lycopene beta cyclase, chloroplastic/chromoplastic</span></span>
<span class="line"><span>  attcgacttatctgggatcttgt</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="_5-column-feature-table" tabindex="-1"><a class="header-anchor" href="#_5-column-feature-table"><span>5-Column Feature Table</span></a></h3><p>How can you generate a 5-column feature table from GenBank format?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  XtoT() {</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    xtract -pattern INSDSeq -pfx &quot;&gt;Feature &quot; \\</span></span>
<span class="line"><span>        -first INSDSeqid,INSDSeq_accession-version \\</span></span>
<span class="line"><span>      -group INSDFeature -FKEY INSDFeature_key \\</span></span>
<span class="line"><span>        -block INSDInterval -deq &quot;\\n&quot; \\</span></span>
<span class="line"><span>          -element INSDInterval_from INSDInterval_to \\</span></span>
<span class="line"><span>            INSDInterval_point INSDInterval_point \\</span></span>
<span class="line"><span>            &quot;&amp;FKEY&quot; -FKEY &quot;()&quot; \\</span></span>
<span class="line"><span>        -block INSDQualifier -deq &quot;\\n\\t\\t\\t&quot; \\</span></span>
<span class="line"><span>          -element INSDQualifier_name INSDQualifier_value</span></span>
<span class="line"><span>  }</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  efetch -db nuccore -id U54469 -format gb |</span></span>
<span class="line"><span>  transmute -g2x |</span></span>
<span class="line"><span>  XtoT</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Exploring the INSDSeq XML hierarchy with a ‑pattern {sequence} ‑group {feature} ‑block {qualifier} construct, and adding proper indentation, can produce feature table submission format:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &gt;Feature U54469.1</span></span>
<span class="line"><span>  1       2881    source</span></span>
<span class="line"><span>                            organism      Drosophila melanogaster</span></span>
<span class="line"><span>                            mol_type      genomic DNA</span></span>
<span class="line"><span>                            db_xref       taxon:7227</span></span>
<span class="line"><span>                            chromosome    3</span></span>
<span class="line"><span>                            map           67A8-B2</span></span>
<span class="line"><span>  80      2881    gene</span></span>
<span class="line"><span>                            gene          eIF4E</span></span>
<span class="line"><span>  80      224     mRNA</span></span>
<span class="line"><span>  892     1458</span></span>
<span class="line"><span>  1550    1920</span></span>
<span class="line"><span>  1986    2085</span></span>
<span class="line"><span>  2317    2404</span></span>
<span class="line"><span>  2466    2881</span></span>
<span class="line"><span>                            gene          eIF4E</span></span>
<span class="line"><span>                            product       eukaryotic initiation factor 4E-I</span></span>
<span class="line"><span>                            ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>An <strong>xml2tbl</strong> script containing this xtract command is now included with EDirect.</p><h3 id="wgs-components" tabindex="-1"><a class="header-anchor" href="#wgs-components"><span>WGS Components</span></a></h3><p>How can you get FASTA format for all components of a WGS project?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  GenerateWGSAccessions() {</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    gb=$( efetch -db nuccore -id &quot;$1&quot; -format gbc &lt; /dev/null )</span></span>
<span class="line"><span>    ln=$( echo &quot;$gb&quot; | xtract -pattern INSDSeq -element INSDSeq_length )</span></span>
<span class="line"><span>    pf=$(</span></span>
<span class="line"><span>      echo &quot;$gb&quot; |</span></span>
<span class="line"><span>      xtract -pattern INSDSeq -element INSDSeq_locus |</span></span>
<span class="line"><span>      sed -e &#39;s/010*/01/g&#39;</span></span>
<span class="line"><span>    )</span></span>
<span class="line"><span>    seq -f &quot;\${pf}%07.0f&quot; 1 &quot;$ln&quot;</span></span>
<span class="line"><span>  }</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  GenerateWGSAccessions &quot;JABRPF000000000&quot; |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>returns the expanded list of accessions from the WGS master:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  JABRPF010000001</span></span>
<span class="line"><span>  JABRPF010000002</span></span>
<span class="line"><span>  JABRPF010000003</span></span>
<span class="line"><span>  ...</span></span>
<span class="line"><span>  JABRPF010000061</span></span>
<span class="line"><span>  JABRPF010000062</span></span>
<span class="line"><span>  JABRPF010000063</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Piping those accessions to:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -format fasta</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>retrieves the individual components in FASTA format:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &gt;JABRPF010000001.1 Enterococcus faecalis strain G109-2 contig00001, ...</span></span>
<span class="line"><span>  ACACTAATATGTGTCTTTTTAGACACTAGCTCACTAAAAAAAATAGTCATAATTTCTTCATTATTAAAAT</span></span>
<span class="line"><span>  CCAACAATTGTGAAATCAATTTAATATCCGATGCTTTGAAAACAACTTCTCCTTTTAATTTTTTGTAAAT</span></span>
<span class="line"><span>  CGTTGAAGCGGATATTGGTTGCCCGTATGCAGTCATTTCATTTGCCAACCACTCAACATTTTTTCCTTTC</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="proteins-in-a-region" tabindex="-1"><a class="header-anchor" href="#proteins-in-a-region"><span>Proteins in a Region</span></a></h3><p>How do you get the protein records encoded in a specific region of a nucleotide sequence?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id NC_015162.1 -format gb -seq_start 9125 -seq_stop 103803 |</span></span>
<span class="line"><span>  xtract -insd CDS protein_id |</span></span>
<span class="line"><span>  cut -f 2 |</span></span>
<span class="line"><span>  efetch -db protein -format fasta</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>extracts the protein_id accessions and retrieves those proteins in FASTA format:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &gt;WP_013615770.1 hypothetical protein [Deinococcus proteolyticus]</span></span>
<span class="line"><span>  MTQNASAGLTAETLMTETGLSAAAVRRALKAYDAAFGLEHPTQDGELHLTPAEYDVLRRALQLTGGYAPG</span></span>
<span class="line"><span>  LKLWFGEQQALALSTQPVASPREVQQLSQLYQQVESYRARPPEEPAQTLRALLDQAQALGWSGFWEMGRL</span></span>
<span class="line"><span>  QGAALFVLGVLRFEDGQRAKVSMQTPLSGAEALVLSRGVCALLQRVRTQPGSGQAWSWNEVLLEEVERIL</span></span>
<span class="line"><span>  KDLTE</span></span>
<span class="line"><span>  &gt;WP_013615771.1 hypothetical protein [Deinococcus proteolyticus]</span></span>
<span class="line"><span>  MNKDTNVDTAFAGGFFTTFENGECVFYDADGVCHSVTPDTVAAVVASGTPRDVVALHEDVQGRTDAVAQV</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="reverse-complement-genbank" tabindex="-1"><a class="header-anchor" href="#reverse-complement-genbank"><span>Reverse Complement GenBank</span></a></h3><p>How do you reverse complement a nucleotide sequence in GenBank format?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id J01749 -format gb |</span></span>
<span class="line"><span>  gbf2fsa |</span></span>
<span class="line"><span>  transmute -revcomp |</span></span>
<span class="line"><span>  transmute -fasta -width 50</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>converts GenBank to FASTA, reverse-complements the sequence, and saves it at the selected 50 characters per line:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  TTCTTGAAGACGAAAGGGCCTCGTGATACGCCTATTTTTATAGGTTAATG</span></span>
<span class="line"><span>  TCATGATAATAATGGTTTCTTAGACGTCAGGTGGCACTTTTCGGGGAAAT</span></span>
<span class="line"><span>  GTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATTCAAATATGTA</span></span>
<span class="line"><span>  ...</span></span>
<span class="line"><span>  ACGATGAGCGCATTGTTAGATTTCATACACGGTGCCTGACTGCGTTAGCA</span></span>
<span class="line"><span>  ATTTAACTGTGATAAACTACCGCATTAAAGCTTATCGATGATAAGCTGTC</span></span>
<span class="line"><span>  AAACATGAGAA</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="reverse-complement-ncbi2na" tabindex="-1"><a class="header-anchor" href="#reverse-complement-ncbi2na"><span>Reverse Complement NCBI2NA</span></a></h3><p>How do you reverse complement a 2-bit encoded nucleotide sequence?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id J01749 -format asn |</span></span>
<span class="line"><span>  xtract -pattern Seq-entry \\</span></span>
<span class="line"><span>    -group seq -if ncbi2na \\</span></span>
<span class="line"><span>      -block inst -LEN length -SEQ -ncbi2na ncbi2na \\</span></span>
<span class="line"><span>        -SUB &quot;&amp;SEQ[:&amp;LEN]&quot; -REV -revcomp &quot;&amp;SUB&quot; \\</span></span>
<span class="line"><span>        -sep &quot;\\n&quot; -fasta &quot;&amp;REV&quot;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>expands the NCBI2NA representation to IUPACNA, truncates at the indicated length, reverse-complements that sequence, and saves it at the default 70 characters per line:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  TTCTTGAAGACGAAAGGGCCTCGTGATACGCCTATTTTTATAGGTTAATGTCATGATAATAATGGTTTCT</span></span>
<span class="line"><span>  TAGACGTCAGGTGGCACTTTTCGGGGAAATGTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATT</span></span>
<span class="line"><span>  CAAATATGTATCCGCTCATGAGACAATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAGTAT</span></span>
<span class="line"><span>  ...</span></span>
<span class="line"><span>  GCATAACCAAGCCTATGCCTACAGCATCCAGGGTGACGGTGCCGAGGATGACGATGAGCGCATTGTTAGA</span></span>
<span class="line"><span>  TTTCATACACGGTGCCTGACTGCGTTAGCAATTTAACTGTGATAAACTACCGCATTAAAGCTTATCGATG</span></span>
<span class="line"><span>  ATAAGCTGTCAAACATGAGAA</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="six-frame-protein-translation" tabindex="-1"><a class="header-anchor" href="#six-frame-protein-translation"><span>Six-Frame Protein Translation</span></a></h3><p>How do you translate all six reading frames of a nucleotide sequence?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id J01749 -format fasta |</span></span>
<span class="line"><span>  transmute -cds2prot -circular -gcode 11 -all</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>will simultaneously translate three plus-strand and three minus-strand frames, including codons that span the origin if the molecule is circular:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &gt;J01749.1-1+</span></span>
<span class="line"><span>  FSCLTAYHR*ALMR*FITVKLLTQSGTVYE ... KCHLTSKKPLLS*H*PIKIGVSRGPFVFKN</span></span>
<span class="line"><span>  &gt;J01749.1-2+</span></span>
<span class="line"><span>  SHV*QLIIDKL*CGSLSQLNC*RSQAPCMK ... SAT*RLRNHYYHDINL*K*AYHEALSSSRI</span></span>
<span class="line"><span>  &gt;J01749.1-3+</span></span>
<span class="line"><span>  LMFDSLSSISFNAVVYHS*IANAVRHRV*N ... VPPDV*ETIIIMTLTYKNRRITRPFRLQEF</span></span>
<span class="line"><span>  &gt;J01749.1-1-</span></span>
<span class="line"><span>  ILEDERAS*YAYFYRLMS***WFLRRQVAL ... FIHGA*LR*QFNCDKLPH*SLSMISCQT*E</span></span>
<span class="line"><span>  &gt;J01749.1-2-</span></span>
<span class="line"><span>  NS*RRKGLVIRLFL*VNVMIIMVS*TSGGT ... FHTRCLTALAI*L**TTALKLIDDKLSNMR</span></span>
<span class="line"><span>  &gt;J01749.1-3-</span></span>
<span class="line"><span>  EFLKTKGPRDTPIFIG*CHDNNGFLDVRWH ... ISYTVPDCVSNLTVINYRIKAYR**AVKHE</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="pattern-searching" tabindex="-1"><a class="header-anchor" href="#pattern-searching"><span>Pattern Searching</span></a></h3><p>The pBR322 cloning vector is a circular plasmid with unique restriction sites in two antibiotic resistance genes.</p><p>The transmute -replace function introduces a second BamHI restriction enzyme recognition site by modifying the ribosomal binding site of the pBR322 rop (restrictor of plasmid copy number) gene:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id J01749 -format fasta |</span></span>
<span class="line"><span>  transmute -replace -offset 1907 -delete GG -insert TC |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>The transmute -search function takes a list of sequence patterns with optional labels (such as restriction enzyme names), and uses a finite-state algorithm to simultaneously search for all patterns:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  transmute -search -circular GGATCC:BamHI GAATTC:EcoRI CTGCAG:PstI |</span></span>
<span class="line"><span>  align-columns -g 4 -a rl</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>The (0-based) starting positions and labels for each match are then printed in a two-column table:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>   374    BamHI</span></span>
<span class="line"><span>  1904    BamHI</span></span>
<span class="line"><span>  3606    PstI</span></span>
<span class="line"><span>  4358    EcoRI</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>The <strong>disambiguate-nucleotides</strong> and <strong>systematic-mutations</strong> scripts can generate all possible single-base substitutions in a pattern for a more relaxed search.</p><h2 id="protein" tabindex="-1"><a class="header-anchor" href="#protein"><span>Protein</span></a></h2><h3 id="amino-acid-composition" tabindex="-1"><a class="header-anchor" href="#amino-acid-composition"><span>Amino Acid Composition</span></a></h3><p>What is the amino acid composition of human titin?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  #!/bin/bash -norc</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  AAComp() {</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    abbrev=( Ala Asx Cys Asp Glu Phe Gly His Ile \\</span></span>
<span class="line"><span>             Xle Lys Leu Met Asn Pyl Pro Gln Arg \\</span></span>
<span class="line"><span>             Ser Thr Sec Val Trp Xxx Tyr Glx )</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    tr A-Z a-z |</span></span>
<span class="line"><span>    sed &#39;s/[^a-z]//g&#39; |</span></span>
<span class="line"><span>    fold -w 1 |</span></span>
<span class="line"><span>    sort-uniq-count |</span></span>
<span class="line"><span>    while read num lttr</span></span>
<span class="line"><span>    do</span></span>
<span class="line"><span>      idx=$(printf %i &quot;&#39;$lttr&#39;&quot;)</span></span>
<span class="line"><span>      ofs=$((idx-97)) </span></span>
<span class="line"><span>      echo -e &quot;\${abbrev[$ofs]}\\t$num&quot;</span></span>
<span class="line"><span>    done |</span></span>
<span class="line"><span>    sort</span></span>
<span class="line"><span>  }</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  efetch -db protein -id Q8WZ42 -format gpc |</span></span>
<span class="line"><span>  xtract -pattern INSDSeq -element INSDSeq_sequence |</span></span>
<span class="line"><span>  AAComp</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This produces a table of residue counts using the three-letter amino acid abbreviations:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  Ala    2084</span></span>
<span class="line"><span>  Arg    1640</span></span>
<span class="line"><span>  Asn    1111</span></span>
<span class="line"><span>  Asp    1720</span></span>
<span class="line"><span>  Cys    513</span></span>
<span class="line"><span>  Gln    942</span></span>
<span class="line"><span>  Glu    3193</span></span>
<span class="line"><span>  Gly    2066</span></span>
<span class="line"><span>  His    478</span></span>
<span class="line"><span>  Ile    2062</span></span>
<span class="line"><span>  Leu    2117</span></span>
<span class="line"><span>  Lys    2943</span></span>
<span class="line"><span>  Met    398</span></span>
<span class="line"><span>  Phe    908</span></span>
<span class="line"><span>  Pro    2517</span></span>
<span class="line"><span>  Ser    2463</span></span>
<span class="line"><span>  Thr    2546</span></span>
<span class="line"><span>  Trp    466</span></span>
<span class="line"><span>  Tyr    999</span></span>
<span class="line"><span>  Val    3184</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="longest-sequences" tabindex="-1"><a class="header-anchor" href="#longest-sequences"><span>Longest Sequences</span></a></h3><p>What are the longest known insulin precursor molecules?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db protein -query &quot;insulin [PROT]&quot; |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element Caption Slen Title |</span></span>
<span class="line"><span>  grep -v receptor | sort -k 2,2nr | head -n 5 | cut -f 1 |</span></span>
<span class="line"><span>  xargs -n 1 sh -c &#39;efetch -db protein -id &quot;$0&quot; -format gp &gt; &quot;$0&quot;.gpf&#39;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Post-processing excludes the longer &quot;insulin-like receptor&quot; sequences, sorts by sequence length, and saves the GenPept results to individual files named by their sequence accessions, using the right angle bracket (&quot;<strong>&gt;</strong>&quot;) Unix output redirection character:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  EFN61235.gpf</span></span>
<span class="line"><span>  EFN80340.gpf</span></span>
<span class="line"><span>  EGW08477.gpf</span></span>
<span class="line"><span>  EKC18433.gpf</span></span>
<span class="line"><span>  ELK28555.gpf</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="archaea-enzyme" tabindex="-1"><a class="header-anchor" href="#archaea-enzyme"><span>Archaea Enzyme</span></a></h3><p>Which archaebacteria have chloramphenicol acetyltransferase?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db protein -organism archaea \\</span></span>
<span class="line"><span>    -query &quot;chloramphenicol acetyltransferase [PROT]&quot; |</span></span>
<span class="line"><span>  efetch -format gpc |</span></span>
<span class="line"><span>  xtract -pattern INSDSeq -element INSDSeq_organism INSDSeq_definition |</span></span>
<span class="line"><span>  grep -i chloramphenicol | grep -v MULTISPECIES |</span></span>
<span class="line"><span>  cut -f 1 | sort -f | uniq -i</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Filtering on the definition line produces a list of archaeal organism names:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  Euryarchaeota archaeon</span></span>
<span class="line"><span>  Methanobrevibacter arboriphilus</span></span>
<span class="line"><span>  Methanobrevibacter gottschalkii</span></span>
<span class="line"><span>  Methanobrevibacter millerae</span></span>
<span class="line"><span>  Methanobrevibacter oralis</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h2 id="gene" tabindex="-1"><a class="header-anchor" href="#gene"><span>Gene</span></a></h2><h3 id="gene-counts" tabindex="-1"><a class="header-anchor" href="#gene-counts"><span>Gene Counts</span></a></h3><p>How many genes are on each human chromosome?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  for chr in {1..22} X Y MT</span></span>
<span class="line"><span>  do</span></span>
<span class="line"><span>    esearch -db gene -query &quot;Homo sapiens [ORGN] AND $chr [CHR]&quot; |</span></span>
<span class="line"><span>    efilter -status alive -type coding |</span></span>
<span class="line"><span>    efetch -format docsum |</span></span>
<span class="line"><span>    xtract -pattern DocumentSummary -NAME Name \\</span></span>
<span class="line"><span>      -block GenomicInfoType -if ChrLoc -equals &quot;$chr&quot; \\</span></span>
<span class="line"><span>        -tab &quot;\\n&quot; -element ChrLoc,&quot;&amp;NAME&quot; |</span></span>
<span class="line"><span>    sort | uniq | cut -f 1 |</span></span>
<span class="line"><span>    sort-uniq-count-rank |</span></span>
<span class="line"><span>    reorder-columns 2 1</span></span>
<span class="line"><span>  done</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This returns a count of unique protein-coding genes per chromosome:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  1     2011</span></span>
<span class="line"><span>  2     1224</span></span>
<span class="line"><span>  3     1046</span></span>
<span class="line"><span>  4     742</span></span>
<span class="line"><span>  5     854</span></span>
<span class="line"><span>  6     1015</span></span>
<span class="line"><span>  7     911</span></span>
<span class="line"><span>  8     666</span></span>
<span class="line"><span>  9     763</span></span>
<span class="line"><span>  10    718</span></span>
<span class="line"><span>  11    1279</span></span>
<span class="line"><span>  12    1012</span></span>
<span class="line"><span>  13    327</span></span>
<span class="line"><span>  14    601</span></span>
<span class="line"><span>  15    585</span></span>
<span class="line"><span>  16    835</span></span>
<span class="line"><span>  17    1147</span></span>
<span class="line"><span>  18    266</span></span>
<span class="line"><span>  19    1391</span></span>
<span class="line"><span>  20    528</span></span>
<span class="line"><span>  21    232</span></span>
<span class="line"><span>  22    429</span></span>
<span class="line"><span>  X     839</span></span>
<span class="line"><span>  Y     63</span></span>
<span class="line"><span>  MT    13</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>The range construct cannot be used for Roman numerals, so the equivalent query on <em>Saccharomyces cerevisiae</em> would need to explicitly list all chromosomes, including the mitochondrion:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  for chr in I II III IV V VI VII VIII IX X XI XII XIII XIV XV XVI MT</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>Plastid genes can be selected with &quot;source plastid [PROP]&quot; or ‑location plastid.</p><h3 id="chromosome-locations" tabindex="-1"><a class="header-anchor" href="#chromosome-locations"><span>Chromosome Locations</span></a></h3><p>Where are mammalian calmodulin genes located?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db gene -query &quot;calmodulin * [PFN] AND mammalia [ORGN]&quot; |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary \\</span></span>
<span class="line"><span>    -def &quot;-&quot; -element Id Name MapLocation ScientificName</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>The ‑def command adds a dash to prevent missing data from shifting columns in the table:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  801       CALM1     14q32.11     Homo sapiens</span></span>
<span class="line"><span>  808       CALM3     19q13.32     Homo sapiens</span></span>
<span class="line"><span>  805       CALM2     2p21         Homo sapiens</span></span>
<span class="line"><span>  24242     Calm1     6q32         Rattus norvegicus</span></span>
<span class="line"><span>  12313     Calm1     12 E         Mus musculus</span></span>
<span class="line"><span>  50663     Calm2     6q12         Rattus norvegicus</span></span>
<span class="line"><span>  24244     Calm3     1q21         Rattus norvegicus</span></span>
<span class="line"><span>  12315     Calm3     7 9.15 cM    Mus musculus</span></span>
<span class="line"><span>  12314     Calm2     17 E4        Mus musculus</span></span>
<span class="line"><span>  80796     Calm4     13 A1        Mus musculus</span></span>
<span class="line"><span>  617095    CALM1     -            Bos taurus</span></span>
<span class="line"><span>  396838    CALM3     -            Sus scrofa</span></span>
<span class="line"><span>  520277    CALM3     -            Bos taurus</span></span>
<span class="line"><span>  364774    Calml5    17q12.2      Rattus norvegicus</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="exon-counts" tabindex="-1"><a class="header-anchor" href="#exon-counts"><span>Exon Counts</span></a></h3><p>How many exons are in each dystrophin transcript variant?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db gene -query &quot;DMD [GENE] AND human [ORGN]&quot; |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -block GenomicInfoType \\</span></span>
<span class="line"><span>    -tab &quot;\\n&quot; -element ChrAccVer,ChrStart,ChrStop |</span></span>
<span class="line"><span>  xargs -n 3 sh -c &#39;efetch -db nuccore -format gbc \\</span></span>
<span class="line"><span>    -id &quot;$0&quot; -chr_start &quot;$1&quot; -chr_stop &quot;$2&quot;&#39; |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This retrieves an INSDSeq XML subset record for the indicated gene region, which contains a number of alternatively-spliced dystrophin mRNA and CDS features.</p><p>Data extraction computes the number of intervals for each mRNA location (individual exons plus non-adjacent UTRs), and obtains the transcript sequence accession, transcript length, and product name from qualifiers:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -insd complete mRNA &quot;#INSDInterval&quot; \\</span></span>
<span class="line"><span>    transcript_id &quot;%transcription&quot; product |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>Final filtering and sorting:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  grep -i dystrophin |</span></span>
<span class="line"><span>  sed &#39;s/dystrophin, transcript variant //g&#39; |</span></span>
<span class="line"><span>  sort -k 2,2nr -k 4,4nr</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>results in a table of exon counts and transcript lengths:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  NC_000023.11    79    NM_004010.3       14083    Dp427p2</span></span>
<span class="line"><span>  NC_000023.11    79    NM_004009.3       14000    Dp427p1</span></span>
<span class="line"><span>  NC_000023.11    79    NM_004006.3       13992    Dp427m</span></span>
<span class="line"><span>  NC_000023.11    79    NM_000109.4       13854    Dp427c</span></span>
<span class="line"><span>  NC_000023.11    78    XM_006724468.2    13923    X1</span></span>
<span class="line"><span>  NC_000023.11    78    XM_017029328.1    13916    X4</span></span>
<span class="line"><span>  NC_000023.11    78    XM_006724469.3    13897    X2</span></span>
<span class="line"><span>  NC_000023.11    77    XM_006724470.3    13875    X3</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="upstream-sequences" tabindex="-1"><a class="header-anchor" href="#upstream-sequences"><span>Upstream Sequences</span></a></h3><p>What sequences are upstream of phenylalanine hydroxylase genes?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db nuccore -query &quot;U49897 [ACCN]&quot; |</span></span>
<span class="line"><span>  elink -target gene |</span></span>
<span class="line"><span>  elink -target homologene |</span></span>
<span class="line"><span>  elink -target gene |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -if GenomicInfoType -element Id \\</span></span>
<span class="line"><span>    -block GenomicInfoType -element ChrAccVer -1-based ChrStart ChrStop |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This produces a table with 1-based sequence coordinates:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  5053       NC_000012.12    102958441    102836889</span></span>
<span class="line"><span>  18478      NC_000076.7     87357657     87419999</span></span>
<span class="line"><span>  38871      NT_037436.4     7760453      7763166</span></span>
<span class="line"><span>  24616      NC_005106.4     28066639     28129772</span></span>
<span class="line"><span>  378962     NC_007115.7     17409367     17391680</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Then, given a shell script named &quot;upstream.sh&quot;:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  #!/bin/bash -norc</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  bases=1500</span></span>
<span class="line"><span>  if [ -n &quot;$1&quot; ]</span></span>
<span class="line"><span>  then</span></span>
<span class="line"><span>    bases=&quot;$1&quot;</span></span>
<span class="line"><span>  fi</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  while read id accn start stop</span></span>
<span class="line"><span>  do</span></span>
<span class="line"><span>    if [ $start -eq 0 ] || [ $stop -eq 0 ] || [ $start -eq $stop ]</span></span>
<span class="line"><span>    then</span></span>
<span class="line"><span>      echo &quot;Skipping $id due to ambiguous coordinates&quot;</span></span>
<span class="line"><span>      continue</span></span>
<span class="line"><span>    fi</span></span>
<span class="line"><span>    if [ $start -gt $stop ]</span></span>
<span class="line"><span>    then</span></span>
<span class="line"><span>      stop=$(( start + bases ))</span></span>
<span class="line"><span>      start=$(( start + 1 ))</span></span>
<span class="line"><span>      strand=2</span></span>
<span class="line"><span>    else</span></span>
<span class="line"><span>      stop=$(( start - 1 ))</span></span>
<span class="line"><span>      start=$(( start - bases ))</span></span>
<span class="line"><span>      strand=1</span></span>
<span class="line"><span>    fi</span></span>
<span class="line"><span>    rslt=$( efetch -db nuccore -id $accn -format fasta \\</span></span>
<span class="line"><span>            -seq_start $start -seq_stop $stop -strand $strand &lt; /dev/null )</span></span>
<span class="line"><span>    echo &quot;$rslt&quot;</span></span>
<span class="line"><span>  done</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>the data lines can be piped through:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  upstream.sh 500</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>to extract and print the 500 nucleotides immediately upstream of each gene:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &gt;NC_000012.12:c102958941-102958442 Homo sapiens chromosome 12, GRCh38.p13 ...</span></span>
<span class="line"><span>  TGAAGTCGAGAAGCTCCTGCTCCTCGGGGCTGAGCGGGTCGTAAGAGCCCTCGTCCGACGAGTAGGATGA</span></span>
<span class="line"><span>  GACCGGCGAGCCGGCCATGGAGTTCAAGTCGTTGGAGTAGTTGGGGGAGATGGTGGGCGACAGGACGCCT</span></span>
<span class="line"><span>  GCCTGGAAGGCGGCGCTCACCGCGTCATGCTCGTCCAGCAGCTGCTGCAGCGCGCGGATGTACTCGACCG</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h2 id="assembly" tabindex="-1"><a class="header-anchor" href="#assembly"><span>Assembly</span></a></h2><h3 id="complete-genomes" tabindex="-1"><a class="header-anchor" href="#complete-genomes"><span>Complete Genomes</span></a></h3><p>What complete genomes are available for <em>Escherichia coli</em>?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db assembly -query \\</span></span>
<span class="line"><span>    &quot;Escherichia coli [ORGN] AND representative [PROP]&quot; |</span></span>
<span class="line"><span>  elink -target nuccore -name assembly_nuccore_refseq |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element AccessionVersion Slen Title |</span></span>
<span class="line"><span>  sed &#39;s/,.*//&#39; |</span></span>
<span class="line"><span>  sort-table -k 2,2nr</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This search finds genomic assemblies and sorts the results by sequence length, allowing complete genomes to be easily distinguished from smaller plasmids:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  NC_002695.2    5498578    Escherichia coli O157:H7 str. Sakai DNA</span></span>
<span class="line"><span>  NC_000913.3    4641652    Escherichia coli str. K-12 substr. MG1655</span></span>
<span class="line"><span>  NC_002128.1    92721      Escherichia coli O157:H7 str. Sakai plasmid pO157</span></span>
<span class="line"><span>  NC_002127.1    3306       Escherichia coli O157:H7 str. Sakai plasmid pOSAK1</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>The sed command removes text (e.g., complete genome, complete sequence, primary assembly) after a comma.</p><p>A similar query for humans, additionally filtering out scaffolds, contigs, and plasmids:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db assembly -query &quot;Homo sapiens [ORGN] AND representative [PROP]&quot; |</span></span>
<span class="line"><span>  elink -target nuccore -name assembly_nuccore_refseq |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element AccessionVersion Slen Title |</span></span>
<span class="line"><span>  sed &#39;s/,.*//&#39; |</span></span>
<span class="line"><span>  grep -v scaffold | grep -v contig | grep -v plasmid | grep -v patch |</span></span>
<span class="line"><span>  sort</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>returns the assembled chromosome and mitochondrial sequence records:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  NC_000001.11    248956422    Homo sapiens chromosome 1</span></span>
<span class="line"><span>  NC_000002.12    242193529    Homo sapiens chromosome 2</span></span>
<span class="line"><span>  NC_000003.12    198295559    Homo sapiens chromosome 3</span></span>
<span class="line"><span>  NC_000004.12    190214555    Homo sapiens chromosome 4</span></span>
<span class="line"><span>  NC_000005.10    181538259    Homo sapiens chromosome 5</span></span>
<span class="line"><span>  NC_000006.12    170805979    Homo sapiens chromosome 6</span></span>
<span class="line"><span>  NC_000007.14    159345973    Homo sapiens chromosome 7</span></span>
<span class="line"><span>  NC_000008.11    145138636    Homo sapiens chromosome 8</span></span>
<span class="line"><span>  NC_000009.12    138394717    Homo sapiens chromosome 9</span></span>
<span class="line"><span>  NC_000010.11    133797422    Homo sapiens chromosome 10</span></span>
<span class="line"><span>  NC_000011.10    135086622    Homo sapiens chromosome 11</span></span>
<span class="line"><span>  NC_000012.12    133275309    Homo sapiens chromosome 12</span></span>
<span class="line"><span>  NC_000013.11    114364328    Homo sapiens chromosome 13</span></span>
<span class="line"><span>  NC_000014.9     107043718    Homo sapiens chromosome 14</span></span>
<span class="line"><span>  NC_000015.10    101991189    Homo sapiens chromosome 15</span></span>
<span class="line"><span>  NC_000016.10    90338345     Homo sapiens chromosome 16</span></span>
<span class="line"><span>  NC_000017.11    83257441     Homo sapiens chromosome 17</span></span>
<span class="line"><span>  NC_000018.10    80373285     Homo sapiens chromosome 18</span></span>
<span class="line"><span>  NC_000019.10    58617616     Homo sapiens chromosome 19</span></span>
<span class="line"><span>  NC_000020.11    64444167     Homo sapiens chromosome 20</span></span>
<span class="line"><span>  NC_000021.9     46709983     Homo sapiens chromosome 21</span></span>
<span class="line"><span>  NC_000022.11    50818468     Homo sapiens chromosome 22</span></span>
<span class="line"><span>  NC_000023.11    156040895    Homo sapiens chromosome X</span></span>
<span class="line"><span>  NC_000024.10    57227415     Homo sapiens chromosome Y</span></span>
<span class="line"><span>  NC_012920.1     16569        Homo sapiens mitochondrion</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This process can be automated to loop through a list of specified organisms:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  for org in \\</span></span>
<span class="line"><span>    &quot;Agrobacterium tumefaciens&quot; \\</span></span>
<span class="line"><span>    &quot;Bacillus anthracis&quot; \\</span></span>
<span class="line"><span>    &quot;Escherichia coli&quot; \\</span></span>
<span class="line"><span>    &quot;Neisseria gonorrhoeae&quot; \\</span></span>
<span class="line"><span>    &quot;Pseudomonas aeruginosa&quot; \\</span></span>
<span class="line"><span>    &quot;Shigella flexneri&quot; \\</span></span>
<span class="line"><span>    &quot;Streptococcus pneumoniae&quot;</span></span>
<span class="line"><span>  do</span></span>
<span class="line"><span>    esearch -db assembly -query &quot;$org [ORGN]&quot; |</span></span>
<span class="line"><span>    efilter -query &quot;representative [PROP]&quot; |</span></span>
<span class="line"><span>    elink -target nuccore -name assembly_nuccore_refseq |</span></span>
<span class="line"><span>    efetch -format docsum |</span></span>
<span class="line"><span>    xtract -pattern DocumentSummary -element AccessionVersion Slen Title |</span></span>
<span class="line"><span>    sed &#39;s/,.*//&#39; |</span></span>
<span class="line"><span>    grep -v -i -e scaffold -e contig -e plasmid -e sequence -e patch |</span></span>
<span class="line"><span>    sort-table -k 2,2nr</span></span>
<span class="line"><span>  done</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>which generates:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  NC_011985.1    4005130    Agrobacterium radiobacter K84 chromosome 1</span></span>
<span class="line"><span>  NC_011983.1    2650913    Agrobacterium radiobacter K84 chromosome 2</span></span>
<span class="line"><span>  NC_005945.1    5228663    Bacillus anthracis str. Sterne chromosome</span></span>
<span class="line"><span>  NC_003997.3    5227293    Bacillus anthracis str. Ames chromosome</span></span>
<span class="line"><span>  NC_002695.1    5498450    Escherichia coli O157:H7 str. Sakai chromosome</span></span>
<span class="line"><span>  NC_018658.1    5273097    Escherichia coli O104:H4 str. 2011C-3493 ...</span></span>
<span class="line"><span>  NC_011751.1    5202090    Escherichia coli UMN026 chromosome</span></span>
<span class="line"><span>  NC_011750.1    5132068    Escherichia coli IAI39 chromosome</span></span>
<span class="line"><span>  NC_017634.1    4747819    Escherichia coli O83:H1 str. NRG 857C chromosome</span></span>
<span class="line"><span>  NC_000913.3    4641652    Escherichia coli str. K-12 substr. MG1655</span></span>
<span class="line"><span>  NC_002946.2    2153922    Neisseria gonorrhoeae FA 1090 chromosome</span></span>
<span class="line"><span>  NC_002516.2    6264404    Pseudomonas aeruginosa PAO1 chromosome</span></span>
<span class="line"><span>  NC_004337.2    4607202    Shigella flexneri 2a str. 301 chromosome</span></span>
<span class="line"><span>  NC_003028.3    2160842    Streptococcus pneumoniae TIGR4 chromosome</span></span>
<span class="line"><span>  NC_003098.1    2038615    Streptococcus pneumoniae R6 chromosome</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h2 id="snp" tabindex="-1"><a class="header-anchor" href="#snp"><span>SNP</span></a></h2><h3 id="snp-data-table" tabindex="-1"><a class="header-anchor" href="#snp-data-table"><span>SNP Data Table</span></a></h3><p>How can you obtain a tab-delimited table of SNP attributes from a document summary?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db snp -id 11549407 -format docsum |</span></span>
<span class="line"><span>  snp2tbl</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>The <strong>snp2tbl</strong> script extracts fields from HGVS data and prints the individual values for further processing:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  rs11549407    NC_000011.10    5226773    G    A    Genomic    Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NC_000011.10    5226773    G    C    Genomic    Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NC_000011.10    5226773    G    T    Genomic    Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NG_000007.3     70841      C    A    Genomic    Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NG_000007.3     70841      C    G    Genomic    Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NG_000007.3     70841      C    T    Genomic    Substitution    HBB</span></span>
<span class="line"><span>  ... </span></span>
<span class="line"><span>  rs11549407    NM_000518.5     167        C    A    Coding     Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NM_000518.5     167        C    G    Coding     Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NM_000518.5     167        C    T    Coding     Substitution    HBB</span></span>
<span class="line"><span>  rs11549407    NP_000509.1     39         Q    *    Protein    Termination     HBB</span></span>
<span class="line"><span>  rs11549407    NP_000509.1     39         Q    E    Protein    Missense        HBB</span></span>
<span class="line"><span>  rs11549407    NP_000509.1     39         Q    K    Protein    Missense        HBB</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Columns are SNP identifier, accession.version, offset from start of sequence, letters to delete, letters to insert, sequence class, variant type, and gene name. The internal steps (<strong>snp2hgvs</strong>, <strong>hgvs2spdi</strong>, and <strong>spdi2tbl</strong>) are described below.</p><h3 id="amino-acid-substitutions" tabindex="-1"><a class="header-anchor" href="#amino-acid-substitutions"><span>Amino Acid Substitutions</span></a></h3><p>What are the missense products of green-sensitive opsin?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db gene -query &quot;OPN1MW [PREF] AND human [ORGN]&quot; |</span></span>
<span class="line"><span>  elink -target snp | efilter -class missense |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>SNP document summaries contain HGVS data of the form:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  NC_000023.11:g.154193517C&gt;A, ... ,NP_000504.1:p.Ala285Val</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>This can be parsed by an xtract <strong>‑hgvs</strong> command within the <strong>snp2hgvs</strong> script:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  snp2hgvs |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>into a structured representation of nucleotide and amino acid replacements:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  ...</span></span>
<span class="line"><span>  &lt;HGVS&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;782327292&lt;/Id&gt;</span></span>
<span class="line"><span>    &lt;Gene&gt;OPN1MW&lt;/Gene&gt;</span></span>
<span class="line"><span>    &lt;Variant&gt;</span></span>
<span class="line"><span>      &lt;Class&gt;Genomic&lt;/Class&gt;</span></span>
<span class="line"><span>      &lt;Type&gt;Substitution&lt;/Type&gt;</span></span>
<span class="line"><span>      &lt;Accession&gt;NC_000023.11&lt;/Accession&gt;</span></span>
<span class="line"><span>      &lt;Position&gt;154193516&lt;/Position&gt;</span></span>
<span class="line"><span>      &lt;Deleted&gt;C&lt;/Deleted&gt;</span></span>
<span class="line"><span>      &lt;Inserted&gt;A&lt;/Inserted&gt;</span></span>
<span class="line"><span>      &lt;Hgvs&gt;NC_000023.11:g.154193517C&amp;gt;A&lt;/Hgvs&gt;</span></span>
<span class="line"><span>    &lt;/Variant&gt;</span></span>
<span class="line"><span>    ...</span></span>
<span class="line"><span>    &lt;Variant&gt;</span></span>
<span class="line"><span>      &lt;Class&gt;Coding&lt;/Class&gt;</span></span>
<span class="line"><span>      &lt;Type&gt;Substitution&lt;/Type&gt;</span></span>
<span class="line"><span>      &lt;Accession&gt;NM_000513.2&lt;/Accession&gt;</span></span>
<span class="line"><span>      &lt;Offset&gt;853&lt;/Offset&gt;</span></span>
<span class="line"><span>      &lt;Deleted&gt;C&lt;/Deleted&gt;</span></span>
<span class="line"><span>      &lt;Inserted&gt;T&lt;/Inserted&gt;</span></span>
<span class="line"><span>      &lt;Hgvs&gt;NM_000513.2:c.854C&amp;gt;T&lt;/Hgvs&gt;</span></span>
<span class="line"><span>    &lt;/Variant&gt;</span></span>
<span class="line"><span>    &lt;Variant&gt;</span></span>
<span class="line"><span>      &lt;Class&gt;Protein&lt;/Class&gt;</span></span>
<span class="line"><span>      &lt;Type&gt;Missense&lt;/Type&gt;</span></span>
<span class="line"><span>      &lt;Accession&gt;NP_000504.1&lt;/Accession&gt;</span></span>
<span class="line"><span>      &lt;Position&gt;284&lt;/Position&gt;</span></span>
<span class="line"><span>      &lt;Deleted&gt;A&lt;/Deleted&gt;</span></span>
<span class="line"><span>      &lt;Inserted&gt;D&lt;/Inserted&gt;</span></span>
<span class="line"><span>      &lt;Hgvs&gt;NP_000504.1:p.Ala285Asp&lt;/Hgvs&gt;</span></span>
<span class="line"><span>    &lt;/Variant&gt;</span></span>
<span class="line"><span>    &lt;Variant&gt;</span></span>
<span class="line"><span>      &lt;Class&gt;Protein&lt;/Class&gt;</span></span>
<span class="line"><span>      &lt;Type&gt;Missense&lt;/Type&gt;</span></span>
<span class="line"><span>      &lt;Accession&gt;NP_000504.1&lt;/Accession&gt;</span></span>
<span class="line"><span>      &lt;Position&gt;284&lt;/Position&gt;</span></span>
<span class="line"><span>      &lt;Deleted&gt;A&lt;/Deleted&gt;</span></span>
<span class="line"><span>      &lt;Inserted&gt;V&lt;/Inserted&gt;</span></span>
<span class="line"><span>      &lt;Hgvs&gt;NP_000504.1:p.Ala285Val&lt;/Hgvs&gt;</span></span>
<span class="line"><span>    &lt;/Variant&gt;</span></span>
<span class="line"><span>  &lt;/HGVS&gt;</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>where the original 1-based HGVS positions are converted to 0-based in the XML.</p><p>Passing those results through the <strong>hgvs2spdi</strong> script:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  hgvs2spdi |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>converts CDS-relative offsets to sequence-relative positions:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  ...</span></span>
<span class="line"><span>  &lt;SPDI&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;782327292&lt;/Id&gt;</span></span>
<span class="line"><span>    &lt;Gene&gt;OPN1MW&lt;/Gene&gt;</span></span>
<span class="line"><span>    ...</span></span>
<span class="line"><span>    &lt;Variant&gt;</span></span>
<span class="line"><span>      &lt;Class&gt;Coding&lt;/Class&gt;</span></span>
<span class="line"><span>      &lt;Type&gt;Substitution&lt;/Type&gt;</span></span>
<span class="line"><span>      &lt;Accession&gt;NM_000513.2&lt;/Accession&gt;</span></span>
<span class="line"><span>      &lt;Position&gt;935&lt;/Position&gt;</span></span>
<span class="line"><span>      &lt;Offset&gt;853&lt;/Offset&gt;</span></span>
<span class="line"><span>      &lt;Deleted&gt;C&lt;/Deleted&gt;</span></span>
<span class="line"><span>      &lt;Inserted&gt;A&lt;/Inserted&gt;</span></span>
<span class="line"><span>      &lt;Hgvs&gt;NM_000513.2:c.854C&amp;gt;A&lt;/Hgvs&gt;</span></span>
<span class="line"><span>      &lt;Spdi&gt;NM_000513.2:935:C:A&lt;/Spdi&gt;</span></span>
<span class="line"><span>    &lt;/Variant&gt;</span></span>
<span class="line"><span>    &lt;Variant&gt;</span></span>
<span class="line"><span>      &lt;Class&gt;Coding&lt;/Class&gt;</span></span>
<span class="line"><span>      &lt;Type&gt;Substitution&lt;/Type&gt;</span></span>
<span class="line"><span>      &lt;Accession&gt;NM_000513.2&lt;/Accession&gt;</span></span>
<span class="line"><span>      &lt;Position&gt;935&lt;/Position&gt;</span></span>
<span class="line"><span>      &lt;Offset&gt;853&lt;/Offset&gt;</span></span>
<span class="line"><span>      &lt;Deleted&gt;C&lt;/Deleted&gt;</span></span>
<span class="line"><span>      &lt;Inserted&gt;T&lt;/Inserted&gt;</span></span>
<span class="line"><span>      &lt;Hgvs&gt;NM_000513.2:c.854C&amp;gt;T&lt;/Hgvs&gt;</span></span>
<span class="line"><span>      &lt;Spdi&gt;NM_000513.2:935:C:T&lt;/Spdi&gt;</span></span>
<span class="line"><span>    &lt;/Variant&gt;</span></span>
<span class="line"><span>    ...</span></span>
<span class="line"><span>  &lt;/SPDI&gt;</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Piping this through the <strong>spdi2tbl</strong> script:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  spdi2tbl |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>completes the final step of the <strong>snp2tbl</strong> process to generate a SNP Data Table. Filtering that through:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  grep Protein | grep Missense | cut -f 1-5</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>results in a table of amino acid substitutions sorted by accession, position, and residue:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  rs1238141906    NP_000504.1    40    E    K</span></span>
<span class="line"><span>  rs1189783086    NP_000504.1    42    P    L</span></span>
<span class="line"><span>  rs1257135801    NP_000504.1    45    H    Y</span></span>
<span class="line"><span>  rs1284438666    NP_000504.1    63    V    I</span></span>
<span class="line"><span>  rs1223726997    NP_000504.1    64    I    T</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Those rows are processed in a while loop that caches the current sequence data:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  while read rsid accn ofs del ins</span></span>
<span class="line"><span>  do</span></span>
<span class="line"><span>    if [ &quot;$accn&quot; != &quot;$last&quot; ]</span></span>
<span class="line"><span>    then</span></span>
<span class="line"><span>      seq=$( efetch -db protein -id &quot;$accn&quot; -format gpc &lt; /dev/null |</span></span>
<span class="line"><span>             xtract -pattern INSDSeq -lower INSDSeq_sequence )</span></span>
<span class="line"><span>      last=&quot;$accn&quot;</span></span>
<span class="line"><span>    fi</span></span>
<span class="line"><span>    pos=$((ofs + 1))</span></span>
<span class="line"><span>    echo &quot;&gt;$rsid [$accn $ins@$pos]&quot;</span></span>
<span class="line"><span>    echo &quot;$seq&quot; |</span></span>
<span class="line"><span>    transmute -replace -offset &quot;$ofs&quot; -delete &quot;$del&quot; -insert &quot;$ins&quot; -lower |</span></span>
<span class="line"><span>    fold -w 50</span></span>
<span class="line"><span>  done</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>and uses transmute <strong>‑replace</strong> to generate modified FASTA with substituted residues in upper case:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &gt;rs1238141906 [NP_000504.1 K@41]</span></span>
<span class="line"><span>  maqqwslqrlagrhpqdsyedstqssiftytnsnstrgpfKgpnyhiapr</span></span>
<span class="line"><span>  wvyhltsvwmifvviasvftnglvlaatmkfkklrhplnwilvnlavadl</span></span>
<span class="line"><span>  aetviastisvvnqvygyfvlghpmcvlegytvslcgitglwslaiiswe</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="snp-modified-product-pairs" tabindex="-1"><a class="header-anchor" href="#snp-modified-product-pairs"><span>SNP-Modified Product Pairs</span></a></h3><p>How can you match codon modifications with amino acid substitutions?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db snp -id 11549407 -format docsum |</span></span>
<span class="line"><span>  snp2tbl |</span></span>
<span class="line"><span>  tbl2prod</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>For SNPs that have different substitutions at the same position, the <strong>tbl2prod</strong> script translates coding sequences (after nucleotide modification), and sorts them with protein sequences (after residue replacement), to produce adjacent matching CDS/protein pairs:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  rs11549407    NM_000518.5:167:C:T    MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWT*R...</span></span>
<span class="line"><span>  rs11549407    NP_000509.1:39:Q:*     MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWT*R...</span></span>
<span class="line"><span>  rs11549407    NM_000518.5:167:C:G    MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTER...</span></span>
<span class="line"><span>  rs11549407    NP_000509.1:39:Q:E     MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTER...</span></span>
<span class="line"><span>  rs11549407    NM_000518.5:167:C:A    MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTKR...</span></span>
<span class="line"><span>  rs11549407    NP_000509.1:39:Q:K     MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTKR...</span></span>
<span class="line"><span>  rs11549407    NM_000518.5:167:C:+    MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQR...</span></span>
<span class="line"><span>  rs11549407    NP_000509.1:39:Q:+     MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQR...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>The &quot;<strong>+</strong>&quot; sign indicates the unmodified &quot;wild-type&quot; nucleotide or amino acid.</p><h2 id="structure" tabindex="-1"><a class="header-anchor" href="#structure"><span>Structure</span></a></h2><h3 id="structural-similarity" tabindex="-1"><a class="header-anchor" href="#structural-similarity"><span>Structural Similarity</span></a></h3><p>What archaea structures are similar to snake venom phospholipase?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db structure -query &quot;crotalus [ORGN] AND phospholipase A2&quot; |</span></span>
<span class="line"><span>  elink -related |</span></span>
<span class="line"><span>  efilter -query &quot;archaea [ORGN]&quot; |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary \\</span></span>
<span class="line"><span>    -if PdbClass -equals Hydrolase \\</span></span>
<span class="line"><span>      -element PdbDescr |</span></span>
<span class="line"><span>  sort -f | uniq -i</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Structure neighbors use geometric comparison to find proteins that are too divergent to be detected by sequence similarity with a BLAST search:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  Crystal Structure Of Autoprocessed Form Of Tk-Subtilisin</span></span>
<span class="line"><span>  Crystal Structure Of Ca2 Site Mutant Of Pro-S324a</span></span>
<span class="line"><span>  Crystal Structure Of Ca3 Site Mutant Of Pro-S324a</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h2 id="taxonomy" tabindex="-1"><a class="header-anchor" href="#taxonomy"><span>Taxonomy</span></a></h2><h3 id="taxonomic-lineage" tabindex="-1"><a class="header-anchor" href="#taxonomic-lineage"><span>Taxonomic Lineage</span></a></h3><p>What are the major taxonomic lineage nodes for humans?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db taxonomy -id 9606 -format xml |</span></span>
<span class="line"><span>  xtract -pattern Taxon -first TaxId -tab &quot;\\n&quot; -element ScientificName \\</span></span>
<span class="line"><span>    -block &quot;**/Taxon&quot; \\</span></span>
<span class="line"><span>      -if Rank -is-not &quot;no rank&quot; -and Rank -is-not clade \\</span></span>
<span class="line"><span>        -tab &quot;\\n&quot; -element Rank,ScientificName</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This uses the double star / child construct to recursively explore the data hierarchy:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  9606            Homo sapiens</span></span>
<span class="line"><span>  superkingdom    Eukaryota</span></span>
<span class="line"><span>  kingdom         Metazoa</span></span>
<span class="line"><span>  phylum          Chordata</span></span>
<span class="line"><span>  subphylum       Craniata</span></span>
<span class="line"><span>  superclass      Sarcopterygii</span></span>
<span class="line"><span>  class           Mammalia</span></span>
<span class="line"><span>  superorder      Euarchontoglires</span></span>
<span class="line"><span>  order           Primates</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="taxonomy-search" tabindex="-1"><a class="header-anchor" href="#taxonomy-search"><span>Taxonomy Search</span></a></h3><p>Which organisms contain an annotated RefSeq genome MatK gene?</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db nuccore -query &quot;MatK [GENE] AND NC_0:NC_999999999 [PACC]&quot; |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element TaxId |</span></span>
<span class="line"><span>  sort -n | uniq |</span></span>
<span class="line"><span>  epost -db taxonomy |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element ScientificName |</span></span>
<span class="line"><span>  sort</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>The first query obtains taxonomy UIDs from nucleotide document summaries and uploads them for separate retrieval from the taxonomy database:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  Acidosasa purpurea</span></span>
<span class="line"><span>  Acorus americanus</span></span>
<span class="line"><span>  ...</span></span>
<span class="line"><span>  Zingiber spectabile</span></span>
<span class="line"><span>  Zygnema circumcarinatum</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h2 id="biosample" tabindex="-1"><a class="header-anchor" href="#biosample"><span>BioSample</span></a></h2><p>BioSample document summaries:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esummary -db biosample -id SAMN34375013 |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>use XML attributes to identify data element types:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &lt;Attribute attribute_name=&quot;strain&quot; harmonized_name=&quot;strain&quot; ...&gt;KF24&lt;/Attribute&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>A two-stage xtract pipeline can produce a tab-delimited table, with one column for each selected field. Piping the data to the first command:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -rec BioSampleInfo -pattern DocumentSummary \\</span></span>
<span class="line"><span>    -wrp Accession -element Accession \\</span></span>
<span class="line"><span>    -wrp Title -element DocumentSummary/Title \\</span></span>
<span class="line"><span>    -wrp Link -sep &quot;|&quot; -numeric Links/Link \\</span></span>
<span class="line"><span>    -group Attribute -if @harmonized_name \\</span></span>
<span class="line"><span>      -TAG -lower @harmonized_name -wrp &quot;&amp;TAG&quot; -element Attribute |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>generates an intermediate form, with XML object names taken from the &quot;harmonized_name&quot; attributes:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &lt;BioSampleInfo&gt;</span></span>
<span class="line"><span>    &lt;Accession&gt;SAMN34375013&lt;/Accession&gt;</span></span>
<span class="line"><span>    &lt;Title&gt;Microbe sample from Bacillus subtilis&lt;/Title&gt;</span></span>
<span class="line"><span>    &lt;Link&gt;960711&lt;/Link&gt;</span></span>
<span class="line"><span>    &lt;isolation_source&gt;rhizosphere soil&lt;/isolation_source&gt;</span></span>
<span class="line"><span>    &lt;collection_date&gt;missing&lt;/collection_date&gt;</span></span>
<span class="line"><span>    &lt;geo_loc_name&gt;China: Kaifeng, Henan Province&lt;/geo_loc_name&gt;</span></span>
<span class="line"><span>    &lt;sample_type&gt;bacterial isolate&lt;/sample_type&gt;</span></span>
<span class="line"><span>    &lt;lat_lon&gt;34.14 N 114.05 E&lt;/lat_lon&gt;</span></span>
<span class="line"><span>    &lt;strain&gt;KF24&lt;/strain&gt;</span></span>
<span class="line"><span>  &lt;/BioSampleInfo&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>(A <strong>bsmp2info</strong> script containing this first xtract command is now included with EDirect.)</p><p>Desired fields can then be selected by name in the last line of the second xtract command, with a &quot;-&quot; placeholder printed for any missing values:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -pattern BioSampleInfo -def &quot;-&quot; -first Accession Title Link \\</span></span>
<span class="line"><span>    geo_loc_name isolation_source strain lat_lon</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>Using ‑first instead of ‑element eliminates possible redundant entries during the attribute name transition from &quot;country&quot; to &quot;geo_loc_name&quot;.</p><h2 id="sra" tabindex="-1"><a class="header-anchor" href="#sra"><span>SRA</span></a></h2><h3 id="using-runinfo-format" tabindex="-1"><a class="header-anchor" href="#using-runinfo-format"><span>Using RunInfo Format</span></a></h3><p>SRA data can be retrieved in RunInfo format:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db sra -id SRR6314034 -format runinfo |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>as comma-separated values, with the first line containing the field names:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  Run,ReleaseDate,LoadDate,spots,bases,spots_with_mates,avgLength,...</span></span>
<span class="line"><span>  SRR6314034,2017-11-21 23:27:11,2017-11-21 23:25:38,128,539118,0,...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>Piping to the <strong>csv2xml</strong> script, and using the ‑header flag:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  csv2xml -set Set -rec Rec -header |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>converts the data into XML:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &lt;Set&gt;</span></span>
<span class="line"><span>    &lt;Rec&gt;</span></span>
<span class="line"><span>      &lt;Run&gt;SRR6314034&lt;/Run&gt;</span></span>
<span class="line"><span>      &lt;ReleaseDate&gt;2017-11-21 23:27:11&lt;/ReleaseDate&gt;</span></span>
<span class="line"><span>      &lt;LoadDate&gt;2017-11-21 23:25:38&lt;/LoadDate&gt;</span></span>
<span class="line"><span>      &lt;spots&gt;128&lt;/spots&gt;</span></span>
<span class="line"><span>      &lt;bases&gt;539118&lt;/bases&gt;</span></span>
<span class="line"><span>      &lt;spots_with_mates&gt;0&lt;/spots_with_mates&gt;</span></span>
<span class="line"><span>      &lt;avgLength&gt;4211&lt;/avgLength&gt;</span></span>
<span class="line"><span>      &lt;size_MB&gt;0&lt;/size_MB&gt;</span></span>
<span class="line"><span>      &lt;AssemblyName&gt;&lt;/AssemblyName&gt;</span></span>
<span class="line"><span>      &lt;download_path&gt;...&lt;/download_path&gt;</span></span>
<span class="line"><span>      &lt;Experiment&gt;SRX3413965&lt;/Experiment&gt;</span></span>
<span class="line"><span>      &lt;LibraryName&gt;child&lt;/LibraryName&gt;</span></span>
<span class="line"><span>      &lt;LibraryStrategy&gt;AMPLICON&lt;/LibraryStrategy&gt;</span></span>
<span class="line"><span>      &lt;LibrarySelection&gt;PCR&lt;/LibrarySelection&gt;</span></span>
<span class="line"><span>      &lt;LibrarySource&gt;GENOMIC&lt;/LibrarySource&gt;</span></span>
<span class="line"><span>      &lt;LibraryLayout&gt;SINGLE&lt;/LibraryLayout&gt;</span></span>
<span class="line"><span>      &lt;InsertSize&gt;0&lt;/InsertSize&gt;</span></span>
<span class="line"><span>      &lt;InsertDev&gt;0&lt;/InsertDev&gt;</span></span>
<span class="line"><span>      &lt;Platform&gt;PACBIO_SMRT&lt;/Platform&gt;</span></span>
<span class="line"><span>      &lt;Model&gt;PacBio RS II&lt;/Model&gt;</span></span>
<span class="line"><span>      &lt;SRAStudy&gt;SRP125431&lt;/SRAStudy&gt;</span></span>
<span class="line"><span>      &lt;BioProject&gt;PRJNA418990&lt;/BioProject&gt;</span></span>
<span class="line"><span>      &lt;Study_Pubmed_id&gt;&lt;/Study_Pubmed_id&gt;</span></span>
<span class="line"><span>      &lt;ProjectID&gt;418990&lt;/ProjectID&gt;</span></span>
<span class="line"><span>      &lt;Sample&gt;SRS2707133&lt;/Sample&gt;</span></span>
<span class="line"><span>      &lt;BioSample&gt;SAMN08040264&lt;/BioSample&gt;</span></span>
<span class="line"><span>      &lt;SampleType&gt;simple&lt;/SampleType&gt;</span></span>
<span class="line"><span>      &lt;TaxID&gt;9606&lt;/TaxID&gt;</span></span>
<span class="line"><span>      &lt;ScientificName&gt;Homo sapiens&lt;/ScientificName&gt;</span></span>
<span class="line"><span>      &lt;SampleName&gt;BBS9_del_mother&lt;/SampleName&gt;</span></span>
<span class="line"><span>      &lt;g1k_pop_code&gt;&lt;/g1k_pop_code&gt;</span></span>
<span class="line"><span>      &lt;source&gt;&lt;/source&gt;</span></span>
<span class="line"><span>      &lt;g1k_analysis_group&gt;&lt;/g1k_analysis_group&gt;</span></span>
<span class="line"><span>      &lt;Subject_ID&gt;&lt;/Subject_ID&gt;</span></span>
<span class="line"><span>      &lt;Sex&gt;female&lt;/Sex&gt;</span></span>
<span class="line"><span>      &lt;Disease&gt;&lt;/Disease&gt;</span></span>
<span class="line"><span>      &lt;Tumor&gt;no&lt;/Tumor&gt;</span></span>
<span class="line"><span>      &lt;Affection_Status&gt;&lt;/Affection_Status&gt;</span></span>
<span class="line"><span>      &lt;Analyte_Type&gt;&lt;/Analyte_Type&gt;</span></span>
<span class="line"><span>      &lt;Histological_Type&gt;&lt;/Histological_Type&gt;</span></span>
<span class="line"><span>      &lt;Body_Site&gt;&lt;/Body_Site&gt;</span></span>
<span class="line"><span>      &lt;CenterName&gt;ICAHN SCHOOL OF MEDICINE AT MOUNT SINAI&lt;/CenterName&gt;</span></span>
<span class="line"><span>      &lt;Submission&gt;SRA633054&lt;/Submission&gt;</span></span>
<span class="line"><span>      &lt;dbgap_study_accession&gt;&lt;/dbgap_study_accession&gt;</span></span>
<span class="line"><span>      &lt;Consent&gt;public&lt;/Consent&gt;</span></span>
<span class="line"><span>      &lt;RunHash&gt;19AC4EB8A65D733274756464DCCF65EA&lt;/RunHash&gt;</span></span>
<span class="line"><span>      &lt;ReadHash&gt;AC8F39C51F95D9CD1BD0CBFBB669AD1E&lt;/ReadHash&gt;</span></span>
<span class="line"><span>    &lt;/Rec&gt;</span></span>
<span class="line"><span>  &lt;/Set&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This can then be piped through xtract:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -pattern Rec -def &quot;-&quot; -element Run Experiment BioProject BioSample</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>to retrieve the desired values by field name:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  SRR6314034    SRX3413965    PRJNA418990    SAMN08040264</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h3 id="installation-of-edirect-on-cloud" tabindex="-1"><a class="header-anchor" href="#installation-of-edirect-on-cloud"><span>Installation of EDirect on Cloud</span></a></h3><p>To install the EDirect software, open a terminal window and execute the following command:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  sh -c &quot;$(curl -fsSL https://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)&quot;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>At the end of installation, answer &quot;y&quot; to have the script run the PATH update command to edit your configuration file, so that EDirect programs can be run in subsequent terminal sessions.</p><p>One installation is complete, run:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  export PATH=\${PATH}:\${HOME}/edirect</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>to set the PATH for the current terminal session.</p><h3 id="downloading-blast-software" tabindex="-1"><a class="header-anchor" href="#downloading-blast-software"><span>Downloading BLAST Software</span></a></h3><p>Obtain the Magic-BLAST software if it is not already installed:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  download-ncbi-software magic-blast</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h3 id="preparing-chromosome-files" tabindex="-1"><a class="header-anchor" href="#preparing-chromosome-files"><span>Preparing Chromosome Files</span></a></h3><p>Retrieve the sequence of human chromosome 7 with:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db nuccore -id NC_000007 -format fasta -immediate &gt; NC_000007.fsa</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>Download protein-coding genes on human chromosome 7 in document summary format, and extract the gene ranges into a tab-delimited table:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db gene -query &quot;Homo sapiens [ORGN] AND 7 [CHR]&quot; |</span></span>
<span class="line"><span>  efilter -status alive -type coding | efetch -format docsum |</span></span>
<span class="line"><span>  gene2range &quot;7&quot; &gt; NC_000007.gen</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="sra-gene-analysis" tabindex="-1"><a class="header-anchor" href="#sra-gene-analysis"><span>SRA Gene Analysis</span></a></h3><p>Run a <strong>magicblast</strong> search of SRR6314034 against chromosome 7:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  magicblast -sra SRR6314034 -subject NC_000007.fsa -outfmt asn |</span></span>
<span class="line"><span>  asn2xml &gt; SRR6314034.xml</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>Passing the alignment details through the <strong>blst2tkns</strong> script:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  cat SRR6314034.xml |</span></span>
<span class="line"><span>  blst2tkns |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>tokenizes the alignment parts:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  index          1</span></span>
<span class="line"><span>  score          6268</span></span>
<span class="line"><span>  start          33088037</span></span>
<span class="line"><span>  stop           33167397</span></span>
<span class="line"><span>  strand         plus</span></span>
<span class="line"><span>  match          12</span></span>
<span class="line"><span>  mismatch       1</span></span>
<span class="line"><span>  genomic-ins    1</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Piping these to the <strong>split-at-intron</strong> script:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  split-at-intron |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>detects large genomic insertions:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  6268    1    plus    33088037..33091003,33163798..33167397</span></span>
<span class="line"><span>  4910    2    plus    33089462..33091003,33163799..33167397</span></span>
<span class="line"><span>  1814    3    plus    33164965..33167397</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>These are filtered by minimum score with:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  filter-columns &#39;$1 &gt; 1000&#39; |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>The <strong>fuse-ranges</strong> script then merges overlapping alignments:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  fuse-ranges |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>into a minimal set of extended ranges:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  plus     33088037    33091003    2967</span></span>
<span class="line"><span>  plus     33163798    33167398    3601</span></span>
<span class="line"><span>  plus     33272248    33272278    31</span></span>
<span class="line"><span>  plus     33394238    33394551    314</span></span>
<span class="line"><span>  minus    33088037    33091004    2968</span></span>
<span class="line"><span>  minus    33163798    33167398    3601</span></span>
<span class="line"><span>  minus    33255886    33256012    127</span></span>
<span class="line"><span>  minus    33394326    33394551    226</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Running each range through the <strong>find-in-gene</strong> script:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  while read std min max len</span></span>
<span class="line"><span>  do</span></span>
<span class="line"><span>    cat NC_000007.gen |</span></span>
<span class="line"><span>    find-in-gene &quot;$std&quot; &quot;$min&quot; &quot;$max&quot;</span></span>
<span class="line"><span>  done |</span></span>
<span class="line"><span>  sort -f | uniq -i</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>returns the names of gene(s) that overlap the aligned segments:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  BBS9</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h2 id="pubchem" tabindex="-1"><a class="header-anchor" href="#pubchem"><span>PubChem</span></a></h2><h3 id="pubchem-in-entrez" tabindex="-1"><a class="header-anchor" href="#pubchem-in-entrez"><span>PubChem in Entrez</span></a></h3><p>Entrez can search on the complete synonym of a compound:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db pccompound -query &quot;catechol [CSYN]&quot; |</span></span>
<span class="line"><span>  efetch -format uid</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>to return a small number of closely matching compound identifiers (CIDs):</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  73160</span></span>
<span class="line"><span>  9064</span></span>
<span class="line"><span>  289</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Entrez document summaries for a compound:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  efetch -db pccompound -id 289 -format docsum |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element MolecularFormula IsomericSmiles</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>contains general descriptive information fields for the compound:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  C6H6O2    C1=CC=C(C(=C1)O)O</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><h3 id="power-user-gateway-pug-rest-query-form" tabindex="-1"><a class="header-anchor" href="#power-user-gateway-pug-rest-query-form"><span>Power User Gateway (PUG) REST Query Form</span></a></h3><p>PubChem also supports a RESTful service for more advanced queries. Nquire provides a ‑pugrest URL shortcut. The base form of a search request is is:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugrest [compound|substance|assay] [input] [operation] [output]</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>Searches can use the name of a compound:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugrest compound name catechol cids TXT</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>to obtain the best matching compound identifier:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  289</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>The CID can be used as the input key to obtain a title and description:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugrest compound cid 289 description XML</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>or to retrieve a much more detailed record:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugrest compound cid 289 record XML |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>that includes the canonical or isomeric SMILES codes:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -pattern PC-InfoData \\</span></span>
<span class="line"><span>    -if PC-Urn_label -equals SMILES -and PC-Urn_name -equals Isomeric \\</span></span>
<span class="line"><span>      -element PC-InfoData_value_sval</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="pubchem-chemical-identifiers" tabindex="-1"><a class="header-anchor" href="#pubchem-chemical-identifiers"><span>PubChem Chemical Identifiers</span></a></h3><p>Certain identifier types require POST arguments to encode special symbols:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugrest compound smiles description XML \\</span></span>
<span class="line"><span>    -smiles &quot;C1=CC=C(C(=C1)O)O&quot; |</span></span>
<span class="line"><span>  xtract -pattern InformationList -element Title Description</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This returns the chemical name and description:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  Catechol    Catechol is a benzenediol comprising...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>Other identifier key types that are encoded in separate arguments are shown below:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugrest compound inchi synonyms TXT \\</span></span>
<span class="line"><span>    -inchi &quot;1S/C6H6O2/c7-5-3-1-2-4-6(5)8/h1-4,7-8H&quot;</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  nquire -pugrest compound inchikey cids JSON \\</span></span>
<span class="line"><span>    -inchikey &quot;YCIMNLLNPGFGHC-UHFFFAOYSA-N&quot;</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  nquire -pugrest compound/fastsubstructure/smarts/cids/XML \\</span></span>
<span class="line"><span>    -smarts &quot;[#7]-[#6]-1=[#6]-[#6](C#C)=[#6](-[#6]-[#8])-[#6]=[#6]-1&quot;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>(Nquire ‑inchi will supply the expected &quot;InChI=&quot; prefix if it is missing in the argument string.)</p><h3 id="pug-rest-asynchronous-queries" tabindex="-1"><a class="header-anchor" href="#pug-rest-asynchronous-queries"><span>PUG-REST Asynchronous Queries</span></a></h3><p>Some PUG-REST queries are computationally intensive and run asynchronously:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugrest compound/superstructure/cid/2244/XML |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>The returned <code>&lt;ListKey&gt;</code> token is piped to an nquire ‑pugwait command, which polls the server until the results are available:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  nquire -pugwait</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div></div></div><p>Identifiers are then downloaded and placed directly in an ENTREZ_DIRECT message:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  &lt;ENTREZ_DIRECT&gt;</span></span>
<span class="line"><span>    &lt;Db&gt;pccompound&lt;/Db&gt;</span></span>
<span class="line"><span>    &lt;Count&gt;3750&lt;/Count&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;87&lt;/Id&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;175&lt;/Id&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;176&lt;/Id&gt;</span></span>
<span class="line"><span>    ...</span></span>
<span class="line"><span>    &lt;Id&gt;162400221&lt;/Id&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;162416056&lt;/Id&gt;</span></span>
<span class="line"><span>    &lt;Id&gt;162417911&lt;/Id&gt;</span></span>
<span class="line"><span>  &lt;/ENTREZ_DIRECT&gt;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Support for this form was added when EDirect was redesigned in 2020.</p><h2 id="bibliometrics" tabindex="-1"><a class="header-anchor" href="#bibliometrics"><span>Bibliometrics</span></a></h2><h3 id="reverse-chronological-order" tabindex="-1"><a class="header-anchor" href="#reverse-chronological-order"><span>Reverse Chronological Order</span></a></h3><p>Repackaging an entire document summary and creating a set of keys for sorting:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  esearch -db pubmed -query &quot;tn3 transposition immunity&quot; |</span></span>
<span class="line"><span>  efetch -format docsum |</span></span>
<span class="line"><span>  xtract -rec Rec -pattern DocumentSummary -INDX &quot;+&quot; \\</span></span>
<span class="line"><span>    -group DocumentSummary -pkg SortKeys \\</span></span>
<span class="line"><span>      -unit DocumentSummary -wrp INDX -element &quot;&amp;INDX&quot; \\</span></span>
<span class="line"><span>      -unit PubDate -wrp YEAR -year PubDate \\</span></span>
<span class="line"><span>      -unit PubDate -wrp DATE -date &quot;*&quot; \\</span></span>
<span class="line"><span>      -unit Title -wrp TITL -lower Title \\</span></span>
<span class="line"><span>      -unit Author -position first -wrp FAUT -lower Name \\</span></span>
<span class="line"><span>      -unit Author -position last -wrp LAUT -lower Name \\</span></span>
<span class="line"><span>      -unit Authors -wrp ANUM -num Author/Name \\</span></span>
<span class="line"><span>      -unit DocumentSummary -wrp SIZE -len &quot;*&quot; \\</span></span>
<span class="line"><span>    -group DocumentSummary -pkg DS -element &quot;*&quot; |</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>produces an intermediate structure with separate containers for the sort keys and the original docsum:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  ...</span></span>
<span class="line"><span>  &lt;Rec&gt;</span></span>
<span class="line"><span>    &lt;SortKeys&gt;</span></span>
<span class="line"><span>      &lt;INDX&gt;5&lt;/INDX&gt;</span></span>
<span class="line"><span>      &lt;YEAR&gt;1989&lt;/YEAR&gt;</span></span>
<span class="line"><span>      &lt;DATE&gt;1989/04&lt;/DATE&gt;</span></span>
<span class="line"><span>      &lt;TITL&gt;nucleotide sequences required for tn3 transposition immunity.&lt;/TITL&gt;</span></span>
<span class="line"><span>      &lt;FAUT&gt;kans ja&lt;/FAUT&gt;</span></span>
<span class="line"><span>      &lt;LAUT&gt;casadaban mj&lt;/LAUT&gt;</span></span>
<span class="line"><span>      &lt;ANUM&gt;2&lt;/ANUM&gt;</span></span>
<span class="line"><span>      &lt;SIZE&gt;1695&lt;/SIZE&gt;</span></span>
<span class="line"><span>    &lt;/SortKeys&gt;</span></span>
<span class="line"><span>    &lt;DS&gt;</span></span>
<span class="line"><span>      &lt;DocumentSummary&gt;</span></span>
<span class="line"><span>        &lt;Id&gt;2539356&lt;/Id&gt;</span></span>
<span class="line"><span>        &lt;PubDate&gt;1989 Apr&lt;/PubDate&gt;</span></span>
<span class="line"><span>        &lt;Source&gt;J Bacteriol&lt;/Source&gt;</span></span>
<span class="line"><span>        ...</span></span>
<span class="line"><span>      &lt;/DocumentSummary&gt;</span></span>
<span class="line"><span>    &lt;/DS&gt;</span></span>
<span class="line"><span>  &lt;/Rec&gt;</span></span>
<span class="line"><span>  ...</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>This allows sorting of records by a new tag derived from ‑year:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  xtract -rec Rec -pattern Rec -sort-rev SortKeys/YEAR |</span></span>
<span class="line"><span>  xtract -set DocumentSummarySet -pattern Rec \\</span></span>
<span class="line"><span>    -group DS/DocumentSummary -element &quot;*&quot; |</span></span>
<span class="line"><span>  transmute -format -combine |</span></span>
<span class="line"><span>  xtract -pattern DocumentSummary -element Id PubDate</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>to present publication records with the most recent shown first:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  36257990    2022 Oct 18</span></span>
<span class="line"><span>  28096365    2017 Jan 31</span></span>
<span class="line"><span>  22624153    2012 Feb</span></span>
<span class="line"><span>  21729108    2011 Sep</span></span>
<span class="line"><span>  8595595     1996 Jan</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><h3 id="count-unique-journals" tabindex="-1"><a class="header-anchor" href="#count-unique-journals"><span>Count Unique Journals</span></a></h3><p>Some bibliometric analyses require fetching huge numbers of PubMed records. One such example is counting the number of unique journals publishing papers indexed under a given category.</p><p>A common function consolidates code to execute a query and fetch the records from the EDirect local PubMed archive. The xtract <strong>‑histogram</strong> shortcut acts as an ‑element command followed by a built-in sort-uniq-count. (This can save significant time when the number of results is in the millions.) The line count gives the number of journals publishing at least one paper that satisfies the query:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  CountUniqueJournals() {</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    qry=&quot;$1&quot;</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    # count number of unique journals publishing papers that match a query</span></span>
<span class="line"><span>    phrase-search -db pubmed -query &quot;$qry&quot; | fetch-pubmed |</span></span>
<span class="line"><span>    xtract -pattern PubmedArticle -histogram Journal/ISOAbbreviation |</span></span>
<span class="line"><span>    wc -l | tr -d &#39; &#39;</span></span>
<span class="line"><span>  }</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>Looping through a series of years, and calling this function with and without a filter of interest in the query:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  for year in {2016..2022}</span></span>
<span class="line"><span>  do</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    # want to find percent of journals still using unstructured date</span></span>
<span class="line"><span>    filt=&quot;medline date [PROP]&quot; </span></span>
<span class="line"><span>    base=&quot;journal article [PTYP] AND $year [YEAR]&quot; </span></span>
<span class="line"><span></span></span>
<span class="line"><span>    # subset of journals publishing a paper with filter condition</span></span>
<span class="line"><span>    subs=$( CountUniqueJournals &quot;$base AND $filt&quot; ) </span></span>
<span class="line"><span></span></span>
<span class="line"><span>    # number of unique journals in selected publication types</span></span>
<span class="line"><span>    totl=$( CountUniqueJournals &quot;$base&quot; ) </span></span>
<span class="line"><span></span></span>
<span class="line"><span>    frac=&quot;-&quot;</span></span>
<span class="line"><span>    if [ &quot;$totl&quot; -gt 0 ] </span></span>
<span class="line"><span>    then</span></span>
<span class="line"><span>      # calculate (integer) percentage</span></span>
<span class="line"><span>      frac=$((subs * 100 / totl)) </span></span>
<span class="line"><span>    fi</span></span>
<span class="line"><span></span></span>
<span class="line"><span>    # print year, journal counts, and percent using filter condition</span></span>
<span class="line"><span>    printf &quot;$year\\t$subs\\t$totl\\t$frac\\n&quot;</span></span>
<span class="line"><span></span></span>
<span class="line"><span>  done | align-columns -h 2 -g 4 -a r</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>returns the year, the numbers of unique journals with the filter and in total for that year, and the ratio of the two as a percentage. In this example the last column shows a steady decrease in the percentage of journals providing an unstructured publication date:</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#24292e;--shiki-dark:#abb2bf;--shiki-light-bg:#fff;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes github-light one-dark-pro vp-code"><code><span class="line"><span>  2016    1933    10362    18</span></span>
<span class="line"><span>  2017    1153    10473    11</span></span>
<span class="line"><span>  2018    1071    10321    10</span></span>
<span class="line"><span>  2019     980    10027     9</span></span>
<span class="line"><span>  2020     919    10579     8</span></span>
<span class="line"><span>  2021     950    10947     8</span></span>
<span class="line"><span>  2022     835    10600     7</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p><a href="https://www.ncbi.nlm.nih.gov/books/about/copyright/" target="_blank" rel="noopener noreferrer">Copyright Notice</a></p><p>Bookshelf ID: NBK565821</p>`,362),t=[l];function p(d,r){return a(),n("div",null,t)}const u=s(i,[["render",p],["__file","07-Entrez Direct Examples_EN.html.vue"]]),h=JSON.parse('{"path":"/E-utilities_EN/07-Entrez%20Direct%20Examples_EN.html","title":"Entrez Direct Examples","lang":"zh-CN","frontmatter":{"order":7,"description":"Entrez Direct Examples Jonathan Kans, PhD. Author Information and Affiliations Created: April 23, 2013; Last Update: April 12, 2024. Estimated reading time: 21 minutes Some earl...","head":[["meta",{"property":"og:url","content":"https://vuepress-theme-hope-docs-demo.netlify.app/E-utilities_zh-CN/E-utilities_EN/07-Entrez%20Direct%20Examples_EN.html"}],["meta",{"property":"og:site_name","content":"E-utilities 汉化文档"}],["meta",{"property":"og:title","content":"Entrez Direct Examples"}],["meta",{"property":"og:description","content":"Entrez Direct Examples Jonathan Kans, PhD. Author Information and Affiliations Created: April 23, 2013; Last Update: April 12, 2024. Estimated reading time: 21 minutes Some earl..."}],["meta",{"property":"og:type","content":"article"}],["meta",{"property":"og:locale","content":"zh-CN"}],["meta",{"property":"og:updated_time","content":"2024-07-10T07:54:22.000Z"}],["meta",{"property":"article:author","content":"WhyLIM"}],["meta",{"property":"article:modified_time","content":"2024-07-10T07:54:22.000Z"}],["script",{"type":"application/ld+json"},"{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Article\\",\\"headline\\":\\"Entrez Direct Examples\\",\\"image\\":[\\"\\"],\\"dateModified\\":\\"2024-07-10T07:54:22.000Z\\",\\"author\\":[{\\"@type\\":\\"Person\\",\\"name\\":\\"WhyLIM\\",\\"url\\":\\"https://github.com/WhyLIM\\"}]}"]]},"headers":[{"level":2,"title":"PubMed","slug":"pubmed","link":"#pubmed","children":[{"level":3,"title":"Author Frequency","slug":"author-frequency","link":"#author-frequency","children":[]},{"level":3,"title":"Publication Distribution","slug":"publication-distribution","link":"#publication-distribution","children":[]},{"level":3,"title":"Treatment Locations","slug":"treatment-locations","link":"#treatment-locations","children":[]},{"level":3,"title":"Indexed Date Fields","slug":"indexed-date-fields","link":"#indexed-date-fields","children":[]},{"level":3,"title":"Digital Object Identifiers","slug":"digital-object-identifiers","link":"#digital-object-identifiers","children":[]},{"level":3,"title":"Reference Formatting","slug":"reference-formatting","link":"#reference-formatting","children":[]}]},{"level":2,"title":"Nucleotide","slug":"nucleotide","link":"#nucleotide","children":[{"level":3,"title":"Coding Sequences","slug":"coding-sequences","link":"#coding-sequences","children":[]},{"level":3,"title":"Untranslated Region Sequences","slug":"untranslated-region-sequences","link":"#untranslated-region-sequences","children":[]},{"level":3,"title":"5-Column Feature Table","slug":"_5-column-feature-table","link":"#_5-column-feature-table","children":[]},{"level":3,"title":"WGS Components","slug":"wgs-components","link":"#wgs-components","children":[]},{"level":3,"title":"Proteins in a Region","slug":"proteins-in-a-region","link":"#proteins-in-a-region","children":[]},{"level":3,"title":"Reverse Complement GenBank","slug":"reverse-complement-genbank","link":"#reverse-complement-genbank","children":[]},{"level":3,"title":"Reverse Complement NCBI2NA","slug":"reverse-complement-ncbi2na","link":"#reverse-complement-ncbi2na","children":[]},{"level":3,"title":"Six-Frame Protein Translation","slug":"six-frame-protein-translation","link":"#six-frame-protein-translation","children":[]},{"level":3,"title":"Pattern Searching","slug":"pattern-searching","link":"#pattern-searching","children":[]}]},{"level":2,"title":"Protein","slug":"protein","link":"#protein","children":[{"level":3,"title":"Amino Acid Composition","slug":"amino-acid-composition","link":"#amino-acid-composition","children":[]},{"level":3,"title":"Longest Sequences","slug":"longest-sequences","link":"#longest-sequences","children":[]},{"level":3,"title":"Archaea Enzyme","slug":"archaea-enzyme","link":"#archaea-enzyme","children":[]}]},{"level":2,"title":"Gene","slug":"gene","link":"#gene","children":[{"level":3,"title":"Gene Counts","slug":"gene-counts","link":"#gene-counts","children":[]},{"level":3,"title":"Chromosome Locations","slug":"chromosome-locations","link":"#chromosome-locations","children":[]},{"level":3,"title":"Exon Counts","slug":"exon-counts","link":"#exon-counts","children":[]},{"level":3,"title":"Upstream Sequences","slug":"upstream-sequences","link":"#upstream-sequences","children":[]}]},{"level":2,"title":"Assembly","slug":"assembly","link":"#assembly","children":[{"level":3,"title":"Complete Genomes","slug":"complete-genomes","link":"#complete-genomes","children":[]}]},{"level":2,"title":"SNP","slug":"snp","link":"#snp","children":[{"level":3,"title":"SNP Data Table","slug":"snp-data-table","link":"#snp-data-table","children":[]},{"level":3,"title":"Amino Acid Substitutions","slug":"amino-acid-substitutions","link":"#amino-acid-substitutions","children":[]},{"level":3,"title":"SNP-Modified Product Pairs","slug":"snp-modified-product-pairs","link":"#snp-modified-product-pairs","children":[]}]},{"level":2,"title":"Structure","slug":"structure","link":"#structure","children":[{"level":3,"title":"Structural Similarity","slug":"structural-similarity","link":"#structural-similarity","children":[]}]},{"level":2,"title":"Taxonomy","slug":"taxonomy","link":"#taxonomy","children":[{"level":3,"title":"Taxonomic Lineage","slug":"taxonomic-lineage","link":"#taxonomic-lineage","children":[]},{"level":3,"title":"Taxonomy Search","slug":"taxonomy-search","link":"#taxonomy-search","children":[]}]},{"level":2,"title":"BioSample","slug":"biosample","link":"#biosample","children":[]},{"level":2,"title":"SRA","slug":"sra","link":"#sra","children":[{"level":3,"title":"Using RunInfo Format","slug":"using-runinfo-format","link":"#using-runinfo-format","children":[]},{"level":3,"title":"Installation of EDirect on Cloud","slug":"installation-of-edirect-on-cloud","link":"#installation-of-edirect-on-cloud","children":[]},{"level":3,"title":"Downloading BLAST Software","slug":"downloading-blast-software","link":"#downloading-blast-software","children":[]},{"level":3,"title":"Preparing Chromosome Files","slug":"preparing-chromosome-files","link":"#preparing-chromosome-files","children":[]},{"level":3,"title":"SRA Gene Analysis","slug":"sra-gene-analysis","link":"#sra-gene-analysis","children":[]}]},{"level":2,"title":"PubChem","slug":"pubchem","link":"#pubchem","children":[{"level":3,"title":"PubChem in Entrez","slug":"pubchem-in-entrez","link":"#pubchem-in-entrez","children":[]},{"level":3,"title":"Power User Gateway (PUG) REST Query Form","slug":"power-user-gateway-pug-rest-query-form","link":"#power-user-gateway-pug-rest-query-form","children":[]},{"level":3,"title":"PubChem Chemical Identifiers","slug":"pubchem-chemical-identifiers","link":"#pubchem-chemical-identifiers","children":[]},{"level":3,"title":"PUG-REST Asynchronous Queries","slug":"pug-rest-asynchronous-queries","link":"#pug-rest-asynchronous-queries","children":[]}]},{"level":2,"title":"Bibliometrics","slug":"bibliometrics","link":"#bibliometrics","children":[{"level":3,"title":"Reverse Chronological Order","slug":"reverse-chronological-order","link":"#reverse-chronological-order","children":[]},{"level":3,"title":"Count Unique Journals","slug":"count-unique-journals","link":"#count-unique-journals","children":[]}]}],"git":{"createdTime":1720598062000,"updatedTime":1720598062000,"contributors":[{"name":"Mli-TB","email":"mli.bio@outlook.com","commits":1}]},"readingTime":{"minutes":20.56,"words":6168},"filePathRelative":"E-utilities_EN/07-Entrez Direct Examples_EN.md","localizedDate":"2024年7月10日","autoDesc":true}');export{u as comp,h as data};
